{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import os\n",
    "from src.tools import Prodigal \n",
    "from src.files import FASTAFile\n",
    "from src.reference import Reference\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How should I select the genomes for model training?\n",
    "genome_ids = ['GCF_000005845.2'] # E. coli K-12\n",
    "genome_ids += ['GCF_000009045.1'] # B. subtilis\n",
    "genome_ids += ['GCF_000006765.1'] # P. aeruginosa\n",
    "genome_ids += ['GCF_000195955.2'] # M. tuberculosis\n",
    "\n",
    "names = dict()\n",
    "names['GCF_000005845.2'] = 'E. coli K-12'\n",
    "names['GCF_000009045.1'] = 'B. subtilis 168'\n",
    "names['GCF_000006765.1'] = 'P. aeruginosa'\n",
    "names['GCF_000195955.2'] = 'M. tuberculosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "prodigal = Prodigal()\n",
    "for genome_id in genome_ids:\n",
    "    if not os.path.exists(f'../data/prodigal/{genome_id}_protein.faa'):\n",
    "        prodigal.run(f'../data/ncbi/genomes/{genome_id}_genomic.fna', f'../data/prodigal/{genome_id}_protein.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Want to get the number of proteins in the NCBI reference. \n",
    "prodigal_fasta_files = {genome_id:FASTAFile(path=f'../data/prodigal/{genome_id}_protein.faa') for genome_id in genome_ids}\n",
    "ncbi_fasta_files = {genome_id:FASTAFile(path=f'../data/ncbi/proteins/{genome_id}_protein.faa') for genome_id in genome_ids}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_hits_df = Reference.load(f'../data/compare/{genome_id}_all_hits.csv')\n",
    "top_hits_df = Reference.load(f'../data/compare/{genome_id}_top_hits.csv')\n",
    "\n",
    "ids = top_hits_df[(top_hits_df.category == 'conflict') & (top_hits_df.n_hits > 1)].index\n",
    "ids_to_inspect = all_hits_df[all_hits_df.query_id.isin(ids) & all_hits_df.in_frame].query_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_hits</th>\n",
       "      <th>n_hits_same_strand</th>\n",
       "      <th>n_hits_opposite_strand</th>\n",
       "      <th>n_hits_in_frame</th>\n",
       "      <th>top_hit_feature</th>\n",
       "      <th>top_hit_contig_id</th>\n",
       "      <th>top_hit_product</th>\n",
       "      <th>top_hit_note</th>\n",
       "      <th>top_hit_pseudo</th>\n",
       "      <th>top_hit_locus_tag</th>\n",
       "      <th>...</th>\n",
       "      <th>overlap_stop</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>top_hit_overlap_fraction</th>\n",
       "      <th>query_overlap_fraction</th>\n",
       "      <th>exact_match</th>\n",
       "      <th>in_frame</th>\n",
       "      <th>in_frame_c_terminus</th>\n",
       "      <th>in_frame_n_terminus</th>\n",
       "      <th>category</th>\n",
       "      <th>sequence_identity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NC_000962.3_104</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>repeat_region</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>none</td>\n",
       "      <td>REP-2, len: 1503 nt. REP251, member of REP13E1...</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>104669</td>\n",
       "      <td>957</td>\n",
       "      <td>0.636727</td>\n",
       "      <td>0.996875</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>conflict</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_000962.3_3531</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>mobile_element</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>none</td>\n",
       "      <td>IS1560-2, len: 1568 nt. Possible Insertion seq...</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>3800796</td>\n",
       "      <td>780</td>\n",
       "      <td>0.497449</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>conflict</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_000962.3_3616</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>repeat_region</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>none</td>\n",
       "      <td>This region is a possible MT-complex-specific ...</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>3884917</td>\n",
       "      <td>960</td>\n",
       "      <td>0.699708</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>conflict</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_hits  n_hits_same_strand  n_hits_opposite_strand  \\\n",
       "id                                                                     \n",
       "NC_000962.3_104        2                   2                       0   \n",
       "NC_000962.3_3531       3                   3                       0   \n",
       "NC_000962.3_3616       3                   3                       0   \n",
       "\n",
       "                  n_hits_in_frame top_hit_feature top_hit_contig_id  \\\n",
       "id                                                                    \n",
       "NC_000962.3_104                 1   repeat_region         NC_000962   \n",
       "NC_000962.3_3531                1  mobile_element         NC_000962   \n",
       "NC_000962.3_3616                1   repeat_region         NC_000962   \n",
       "\n",
       "                 top_hit_product  \\\n",
       "id                                 \n",
       "NC_000962.3_104             none   \n",
       "NC_000962.3_3531            none   \n",
       "NC_000962.3_3616            none   \n",
       "\n",
       "                                                       top_hit_note  \\\n",
       "id                                                                    \n",
       "NC_000962.3_104   REP-2, len: 1503 nt. REP251, member of REP13E1...   \n",
       "NC_000962.3_3531  IS1560-2, len: 1568 nt. Possible Insertion seq...   \n",
       "NC_000962.3_3616  This region is a possible MT-complex-specific ...   \n",
       "\n",
       "                  top_hit_pseudo top_hit_locus_tag  ... overlap_stop  \\\n",
       "id                                                  ...                \n",
       "NC_000962.3_104            False              none  ...       104669   \n",
       "NC_000962.3_3531           False              none  ...      3800796   \n",
       "NC_000962.3_3616           False              none  ...      3884917   \n",
       "\n",
       "                 overlap_length  top_hit_overlap_fraction  \\\n",
       "id                                                          \n",
       "NC_000962.3_104             957                  0.636727   \n",
       "NC_000962.3_3531            780                  0.497449   \n",
       "NC_000962.3_3616            960                  0.699708   \n",
       "\n",
       "                  query_overlap_fraction exact_match in_frame  \\\n",
       "id                                                              \n",
       "NC_000962.3_104                 0.996875       False    False   \n",
       "NC_000962.3_3531                1.000000       False    False   \n",
       "NC_000962.3_3616                1.000000       False    False   \n",
       "\n",
       "                 in_frame_c_terminus in_frame_n_terminus  category  \\\n",
       "id                                                                   \n",
       "NC_000962.3_104                False               False  conflict   \n",
       "NC_000962.3_3531               False               False  conflict   \n",
       "NC_000962.3_3616               False               False  conflict   \n",
       "\n",
       "                 sequence_identity  \n",
       "id                                  \n",
       "NC_000962.3_104                0.0  \n",
       "NC_000962.3_3531               0.0  \n",
       "NC_000962.3_3616               0.0  \n",
       "\n",
       "[3 rows x 52 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_hits_df.loc[ids_to_inspect]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_protein_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>NP_214608.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5173</th>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5174</th>\n",
       "      <td>NP_217903.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5175</th>\n",
       "      <td>NP_217904.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5289</th>\n",
       "      <td>NP_217983.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5290</th>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5291</th>\n",
       "      <td>NP_217984.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     subject_protein_id\n",
       "143         NP_214608.1\n",
       "144                none\n",
       "5173               none\n",
       "5174        NP_217903.1\n",
       "5175        NP_217904.1\n",
       "5289        NP_217983.1\n",
       "5290               none\n",
       "5291        NP_217984.1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_hits_df[all_hits_df.query_id.isin(ids_to_inspect)][['subject_protein_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "compare_df = pd.concat([Reference.load(f'../data/compare/{genome_id}_top_hits.csv').assign(genome_id=genome_id) for genome_id in genome_ids])\n",
    "compare_df['name'] = compare_df.genome_id.map(names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_n_cds</th>\n",
       "      <th>prodigal_n_cds</th>\n",
       "      <th>percent_matches</th>\n",
       "      <th>percent_exact_matches</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E. coli K-12</th>\n",
       "      <td>4298</td>\n",
       "      <td>4319</td>\n",
       "      <td>0.953461</td>\n",
       "      <td>0.900208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B. subtilis 168</th>\n",
       "      <td>4237</td>\n",
       "      <td>4226</td>\n",
       "      <td>0.977993</td>\n",
       "      <td>0.88973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P. aeruginosa</th>\n",
       "      <td>5572</td>\n",
       "      <td>5681</td>\n",
       "      <td>0.977997</td>\n",
       "      <td>0.896849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M. tuberculosis</th>\n",
       "      <td>3906</td>\n",
       "      <td>4085</td>\n",
       "      <td>0.92754</td>\n",
       "      <td>0.717013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ncbi_n_cds  prodigal_n_cds percent_matches  \\\n",
       "E. coli K-12           4298            4319        0.953461   \n",
       "B. subtilis 168        4237            4226        0.977993   \n",
       "P. aeruginosa          5572            5681        0.977997   \n",
       "M. tuberculosis        3906            4085         0.92754   \n",
       "\n",
       "                percent_exact_matches  \n",
       "E. coli K-12                 0.900208  \n",
       "B. subtilis 168               0.88973  \n",
       "P. aeruginosa                0.896849  \n",
       "M. tuberculosis              0.717013  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def table_1(compare_df:pd.DataFrame, prodigal_fasta_files:dict, ncbi_fasta_files:dict, path:str=None):\n",
    "\n",
    "    table_df = pd.DataFrame(index=compare_df.name.unique(), columns=['ncbi_n_cds', 'prodigal_n_cds', 'percent_matches', 'percent_exact_matches'])   \n",
    "    table_df['ncbi_n_cds'] = pd.Series({names[genome_id]:len(fasta_file) for genome_id, fasta_file in ncbi_fasta_files.items()})\n",
    "    table_df['prodigal_n_cds'] = pd.Series({names[genome_id]:len(fasta_file) for genome_id, fasta_file in prodigal_fasta_files.items()})\n",
    "    for name, df in compare_df.groupby('name'):\n",
    "        table_df.loc[name, 'percent_matches'] = (df.category == 'match').sum() / len(df)\n",
    "        table_df.loc[name, 'percent_exact_matches'] = df.exact_match.sum() / len(df)\n",
    "    return table_df\n",
    "\n",
    "table_1(compare_df, prodigal_fasta_files, ncbi_fasta_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What statistics do I care about?\n",
    "# (1) Total number of genes in reference. \n",
    "# (2) Total number of predicted genes. \n",
    "# (3) Total number of matched coding sequences (not including pseudogenes).\n",
    "# (4) Number of exact matches. \n",
    "# (5) Number of genes with boundary errors (characterize the nature of the boundary errors later on).\n",
    "# (2) The false discovery rate. \n",
    "# (3) The number of missed genes. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "frugal",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
