{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4b6549f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import os \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as tick \n",
    "import seaborn as sns \n",
    "from utils import load_ncbi_genome_metadata\n",
    "import src.download as download\n",
    "import src.tools as tools\n",
    "from src import get_genome_id\n",
    "import glob\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from src.files import FASTAFile, InterProScanFile\n",
    "from tqdm import tqdm\n",
    "from src.reference import compare, annotate, reverse_complement\n",
    "import src.reference as reference\n",
    "import matplotlib as mpl \n",
    "from sklearn.linear_model import LinearRegression\n",
    "from src.download import NCBI\n",
    "import re \n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import chisquare\n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2\n",
    "\n",
    "get_percent = lambda n, total : f'{100 * n / total:.2f}%' if (total > 0) else '0%'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54e0b393",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83 phyla in the dataset.\n",
      "56 phyla with NCBI reference genomes.\n"
     ]
    }
   ],
   "source": [
    "# This is the genome metadata for all NCBI reference genomes. \n",
    "ncbi_genome_metadata_df = load_ncbi_genome_metadata(genome_metadata_path='../data/dataset/ncbi_genome_metadata.tsv', taxonomy_metadata_path='../data/dataset/ncbi_taxonomy_metadata.tsv') \n",
    "\n",
    "phyla_in_dataset = pd.read_csv('../data/dataset/dataset.csv', index_col=0, usecols=['id', 'lineage']).lineage\n",
    "phyla_in_dataset = np.unique([lineage.split(';')[1] for lineage in phyla_in_dataset[phyla_in_dataset.str.contains(';')]])\n",
    "phyla_in_dataset = np.unique([phylum for phylum in phyla_in_dataset if (phylum != 'environmental samples')])\n",
    "\n",
    "print(len(phyla_in_dataset), 'phyla in the dataset.')\n",
    "print(ncbi_genome_metadata_df.phylum.nunique(), 'phyla with NCBI reference genomes.')\n",
    "\n",
    "genome_metadata_df = ncbi_genome_metadata_df[~ncbi_genome_metadata_df.phylum.isin(phyla_in_dataset)].copy()\n",
    "genome_metadata_df = genome_metadata_df.sort_values('checkm_completeness', ascending=False)\n",
    "genome_metadata_df = genome_metadata_df.drop_duplicates('phylum', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "22fca2e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NCBI.get_genomes: Downloading data for GCF_021057185.1.: 100%|██████████| 20/20 [00:00<00:00, 2768.33it/s]\n"
     ]
    }
   ],
   "source": [
    "ncbi = download.ncbi.NCBI()\n",
    "ncbi.get_genomes(genome_metadata_df.index, include=['gbff', 'genome'], dirs={'genome':'../data/ncbi/genomes', 'gbff':'../data/ncbi/gbffs'})\n",
    "ncbi.cleanup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da299687",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Prodigal on genomes for rare taxa.: 100%|██████████| 20/20 [00:00<00:00, 20641.26it/s]\n",
      "Running InterProScan on Prodigal predictions for rare taxa.: 100%|██████████| 20/20 [00:00<00:00, 52692.26it/s]\n"
     ]
    }
   ],
   "source": [
    "prodigal = tools.Prodigal()\n",
    "for input_path in tqdm(glob.glob('../data/ncbi/genomes/*'), desc='Running Prodigal on genomes for rare taxa.'):\n",
    "    output_path = f'../data/prodigal/{get_genome_id(input_path)}_protein.faa'\n",
    "    if not os.path.exists(output_path):\n",
    "        prodigal.run(input_path, output_path=output_path)\n",
    "\n",
    "interproscan = tools.InterProScan()\n",
    "for input_path in tqdm(glob.glob('../data/prodigal/*'), desc='Running InterProScan on Prodigal predictions for rare taxa.'):\n",
    "    output_path = f'../data/interpro/{get_genome_id(input_path)}_annotation.tsv'\n",
    "    if not os.path.exists(output_path):\n",
    "        interproscan.run(input_path, output_path)\n",
    "\n",
    "# Prodigal sometimes predicts a string of X amino acids (unknown), which coincide with assembly gaps. The\n",
    "# matching proteins are generally marked as partial in the NCBI reference. Need to figure out a way to handle this. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bc0a8b9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing Prodigal predictions with NCBI reference annotations.:   0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing Prodigal predictions with NCBI reference annotations.: 100%|██████████| 20/20 [15:20<00:00, 46.05s/it]\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists('../data/results/results-2/dataset.csv'):\n",
    "    top_hits_df, all_hits_df = list(), list()\n",
    "    for query_path in tqdm(glob.glob('../data/prodigal/*'), desc='Comparing Prodigal predictions with NCBI reference annotations.'):\n",
    "        reference_path = f'../data/ncbi/gbffs/{get_genome_id(query_path)}_genomic.gbff'\n",
    "        top_hits_df_, all_hits_df_ = compare(query_path, reference_path)\n",
    "        top_hits_df += [top_hits_df_]\n",
    "        all_hits_df += [all_hits_df_]\n",
    "    top_hits_df = pd.concat(top_hits_df)\n",
    "    all_hits_df = pd.concat(all_hits_df)\n",
    "    all_hits_df.to_csv('../data/results/results-2/all_hits.csv')\n",
    "\n",
    "    dataset_df = top_hits_df.merge(genome_metadata_df, left_on='genome_id', right_index=True)\n",
    "    dataset_df = dataset_df.rename(columns={'query_seq':'seq'})\n",
    "    dataset_df = dataset_df[dataset_df.seq.apply(len) < 2000].copy()\n",
    "    dataset_df.to_csv('../data/results/results-2/dataset.csv')\n",
    "else:\n",
    "    dataset_df = pd.read_csv('../data/results/results-2/dataset.csv', index_col=0)\n",
    "    all_hits_df = pd.read_csv('../data/results/results-2/all_hits.csv', index_col=0)\n",
    "\n",
    "\n",
    "dataset_df = annotate(dataset_df)\n",
    "all_hits_df = annotate(all_hits_df, prefix='subject')\n",
    "\n",
    "# Exclude genomes on which Prodigal did horrifically (I think I needed to use a different translation table). \n",
    "# Also, one of the excluded genomes belongs to an endosymbiont which is not assigned a phylum.\n",
    "exclude_genome_ids = ['GCF_029854295.1', 'GCF_021057185.1', 'GCF_016097415.1'] \n",
    "dataset_df = dataset_df[~dataset_df.genome_id.isin(exclude_genome_ids)].copy()\n",
    "\n",
    "# Because I did not use a minimum overlap length, want to make sure there are no unreasonably short matches. Minimum overlap length is 84, so seems to be OK.\n",
    "# dataset_df[dataset_df.match].overlap_length.min()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cf020610",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some sanity checks... \n",
    "# (1) For matches where neither sequence is partial, overlap length should be divisible by three (unless there is a frameshift). \n",
    "# (2) For matches where neither sequence is partial, phase_start and phase_stop should both be divisible by three (unless there is a frameshift). \n",
    "# (3) For conflicts where neither sequence is partial, phase_start and phase_stop should be equal.\n",
    "# (3) For CDS conflicts where neither sequence is partial, phase_start and phase_stop should be equal to overlap_length % 3.\n",
    "\n",
    "# (5) All Prodigal nucleotide sequences should be divisible by 3, regardless of whether or not it is partial. \n",
    "\n",
    "overlap_lengths = dataset_df[dataset_df.match & ~dataset_df.top_hit_ribosomal_slippage & (dataset_df.top_hit_partial == '00')].overlap_length\n",
    "phases = dataset_df[dataset_df.match].phase\n",
    "assert np.all(overlap_lengths % 3 == 0), 'All matches with non-partial sequences should have an overlap length divisible by 3.'\n",
    "assert np.all(phases % 3 == 0), 'All translational starts between matches should be in-phase.'\n",
    "\n",
    "# mask = (dataset_df.query_partial == '00') & (dataset_df.top_hit_partial == '00') & ~dataset_df.top_hit_ribosomal_slippage & dataset_df.conflict & (dataset_df.top_hit_feature == 'CDS') & (dataset_df.overlap_type == 'tandem')\n",
    "# overlap_lengths = dataset_df[mask].overlap_length\n",
    "# phase_starts = dataset_df[mask].phase_start\n",
    "# phase_stops = dataset_df[mask].phase_stop\n",
    "# assert np.all(phase_starts == phase_stops), 'Phase difference of translational starts and stops should be the same for all non-partial CDS conflicts.'\n",
    "# # assert np.all(phase_stops % 3 == 0), 'Phase difference of translational starts and stops should correspond to the overlap length.'\n",
    "# mask.sum()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81acdd3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1 % 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fa80aa7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phase</th>\n",
       "      <th>query_overlap</th>\n",
       "      <th>top_hit_overlap</th>\n",
       "      <th>query_strand</th>\n",
       "      <th>top_hit_strand</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_stop</th>\n",
       "      <th>top_hit_start</th>\n",
       "      <th>top_hit_stop</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NZ_JAALLS010000003.1_72</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>75251</td>\n",
       "      <td>75418</td>\n",
       "      <td>74802</td>\n",
       "      <td>75251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_JAYGHR010000001.1_71</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>76646</td>\n",
       "      <td>76768</td>\n",
       "      <td>76068</td>\n",
       "      <td>76646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_JAYGHR010000024.1_11</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>11965</td>\n",
       "      <td>12147</td>\n",
       "      <td>12147</td>\n",
       "      <td>12800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXIU01000180.1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>246</td>\n",
       "      <td>246</td>\n",
       "      <td>974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXIU01000216.1_6</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9125</td>\n",
       "      <td>9232</td>\n",
       "      <td>7596</td>\n",
       "      <td>9125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXIU01000022.1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXIU01000227.1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>372</td>\n",
       "      <td>372</td>\n",
       "      <td>1217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXIU01000004.1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXIU01000088.1_2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>765</td>\n",
       "      <td>881</td>\n",
       "      <td>1</td>\n",
       "      <td>765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP025523.1_119</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>123123</td>\n",
       "      <td>123221</td>\n",
       "      <td>123221</td>\n",
       "      <td>123772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP025523.1_2213</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2175685</td>\n",
       "      <td>2175942</td>\n",
       "      <td>2175269</td>\n",
       "      <td>2175685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP025523.1_2274</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2226163</td>\n",
       "      <td>2226351</td>\n",
       "      <td>2226351</td>\n",
       "      <td>2228477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP025523.1_2282</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2233252</td>\n",
       "      <td>2233542</td>\n",
       "      <td>2233542</td>\n",
       "      <td>2234969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP025523.1_3535</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3480826</td>\n",
       "      <td>3480930</td>\n",
       "      <td>3479753</td>\n",
       "      <td>3480826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP130454.1_189</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>207759</td>\n",
       "      <td>208202</td>\n",
       "      <td>206500</td>\n",
       "      <td>207759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_NIGF01000019.1_29</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>33409</td>\n",
       "      <td>33540</td>\n",
       "      <td>31163</td>\n",
       "      <td>33409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_NIGF01000029.1_21</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>27302</td>\n",
       "      <td>27487</td>\n",
       "      <td>27487</td>\n",
       "      <td>28506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_NIGF01000003.1_115</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>117075</td>\n",
       "      <td>117191</td>\n",
       "      <td>114556</td>\n",
       "      <td>117075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_014960.1_1919</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2147248</td>\n",
       "      <td>2147469</td>\n",
       "      <td>2146220</td>\n",
       "      <td>2147248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP035449.1_1349</th>\n",
       "      <td>2.0</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1601277</td>\n",
       "      <td>1601525</td>\n",
       "      <td>1601525</td>\n",
       "      <td>1602709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP035449.1_1415</th>\n",
       "      <td>2.0</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1683995</td>\n",
       "      <td>1684138</td>\n",
       "      <td>1681068</td>\n",
       "      <td>1683995</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         phase query_overlap top_hit_overlap  query_strand  \\\n",
       "id                                                                           \n",
       "NZ_JAALLS010000003.1_72    2.0            10              01            -1   \n",
       "NZ_JAYGHR010000001.1_71    2.0            10              01             1   \n",
       "NZ_JAYGHR010000024.1_11    2.0            01              10            -1   \n",
       "NZ_QXIU01000180.1_1        2.0            01              10             1   \n",
       "NZ_QXIU01000216.1_6        2.0            10              01             1   \n",
       "NZ_QXIU01000022.1_1        2.0            01              10             1   \n",
       "NZ_QXIU01000227.1_1        2.0            01              10             1   \n",
       "NZ_QXIU01000004.1_1        2.0            01              10             1   \n",
       "NZ_QXIU01000088.1_2        2.0            10              01             1   \n",
       "NZ_AP025523.1_119          2.0            01              10            -1   \n",
       "NZ_AP025523.1_2213         2.0            10              01             1   \n",
       "NZ_AP025523.1_2274         2.0            01              10            -1   \n",
       "NZ_AP025523.1_2282         2.0            01              10            -1   \n",
       "NZ_AP025523.1_3535         2.0            10              01             1   \n",
       "NZ_CP130454.1_189          2.0            10              01             1   \n",
       "NZ_NIGF01000019.1_29       2.0            10              01             1   \n",
       "NZ_NIGF01000029.1_21       2.0            01              10            -1   \n",
       "NZ_NIGF01000003.1_115      2.0            10              01             1   \n",
       "NC_014960.1_1919           2.0            10              01            -1   \n",
       "NZ_AP035449.1_1349         2.0            01              10            -1   \n",
       "NZ_AP035449.1_1415         2.0            10              01             1   \n",
       "\n",
       "                         top_hit_strand  overlap_length  query_start  \\\n",
       "id                                                                     \n",
       "NZ_JAALLS010000003.1_72              -1               1        75251   \n",
       "NZ_JAYGHR010000001.1_71               1               1        76646   \n",
       "NZ_JAYGHR010000024.1_11              -1               1        11965   \n",
       "NZ_QXIU01000180.1_1                   1               1            1   \n",
       "NZ_QXIU01000216.1_6                   1               1         9125   \n",
       "NZ_QXIU01000022.1_1                   1               1            3   \n",
       "NZ_QXIU01000227.1_1                   1               1            1   \n",
       "NZ_QXIU01000004.1_1                   1               1            1   \n",
       "NZ_QXIU01000088.1_2                   1               1          765   \n",
       "NZ_AP025523.1_119                    -1               1       123123   \n",
       "NZ_AP025523.1_2213                    1               1      2175685   \n",
       "NZ_AP025523.1_2274                   -1               1      2226163   \n",
       "NZ_AP025523.1_2282                   -1               1      2233252   \n",
       "NZ_AP025523.1_3535                    1               1      3480826   \n",
       "NZ_CP130454.1_189                     1               1       207759   \n",
       "NZ_NIGF01000019.1_29                  1               1        33409   \n",
       "NZ_NIGF01000029.1_21                 -1               1        27302   \n",
       "NZ_NIGF01000003.1_115                 1               1       117075   \n",
       "NC_014960.1_1919                     -1               1      2147248   \n",
       "NZ_AP035449.1_1349                   -1               1      1601277   \n",
       "NZ_AP035449.1_1415                    1               1      1683995   \n",
       "\n",
       "                         query_stop  top_hit_start  top_hit_stop  \n",
       "id                                                                \n",
       "NZ_JAALLS010000003.1_72       75418          74802         75251  \n",
       "NZ_JAYGHR010000001.1_71       76768          76068         76646  \n",
       "NZ_JAYGHR010000024.1_11       12147          12147         12800  \n",
       "NZ_QXIU01000180.1_1             246            246           974  \n",
       "NZ_QXIU01000216.1_6            9232           7596          9125  \n",
       "NZ_QXIU01000022.1_1             143            143           982  \n",
       "NZ_QXIU01000227.1_1             372            372          1217  \n",
       "NZ_QXIU01000004.1_1              90             90           701  \n",
       "NZ_QXIU01000088.1_2             881              1           765  \n",
       "NZ_AP025523.1_119            123221         123221        123772  \n",
       "NZ_AP025523.1_2213          2175942        2175269       2175685  \n",
       "NZ_AP025523.1_2274          2226351        2226351       2228477  \n",
       "NZ_AP025523.1_2282          2233542        2233542       2234969  \n",
       "NZ_AP025523.1_3535          3480930        3479753       3480826  \n",
       "NZ_CP130454.1_189            208202         206500        207759  \n",
       "NZ_NIGF01000019.1_29          33540          31163         33409  \n",
       "NZ_NIGF01000029.1_21          27487          27487         28506  \n",
       "NZ_NIGF01000003.1_115        117191         114556        117075  \n",
       "NC_014960.1_1919            2147469        2146220       2147248  \n",
       "NZ_AP035449.1_1349          1601525        1601525       1602709  \n",
       "NZ_AP035449.1_1415          1684138        1681068       1683995  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_df[~dataset_df.top_hit_pseudo & (dataset_df.top_hit_feature == 'CDS')].value_counts(['top_hit_partial', 'top_hit_codon_start'])\n",
    "\n",
    "# +2/-2 phase refers to cases where there are 1 + 2n shared bases; +1 and +4 shifts both fall into this category.\n",
    "\n",
    "# There should not be any non-pseudo tandem overlaps where the overlapping terminus is partial.\n",
    "dataset_df[(dataset_df.overlap_length == 1)][['phase', 'query_overlap', 'top_hit_overlap','query_strand', 'top_hit_strand', 'overlap_length', 'query_start', 'query_stop', 'top_hit_start', 'top_hit_stop']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "eefcaabd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "query_start 117075 subject_start 114556\n",
      "839.6666666666666\n",
      "phase 1\n",
      "\n",
      "query_start 1601525 subject_start 1602709\n",
      "394.6666666666667\n",
      "phase 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from src.reference import Reference \n",
    "\n",
    "query_start, _ = Reference._get_translation_start_stop(start=117075, stop=117191, strand=1, adjust_start=False)\n",
    "subject_start, _ = Reference._get_translation_start_stop(start=114556, stop=117075, strand=1, adjust_start=False)\n",
    "print('query_start',  query_start, 'subject_start', subject_start)\n",
    "print(abs(query_start - subject_start) / 3)\n",
    "phase_map = {0:0, 1:2, 2:1}\n",
    "phase = phase_map[abs(query_start - subject_start) % 3]\n",
    "print('phase', phase)\n",
    "print()\n",
    "\n",
    "query_start, _ = Reference._get_translation_start_stop(start=1601277, stop=1601525, strand=-1, adjust_start=False)\n",
    "subject_start, _ = Reference._get_translation_start_stop(start=1601525, stop=1602709, strand=-1, adjust_start=False)\n",
    "print('query_start',  query_start, 'subject_start', subject_start)\n",
    "print(abs(query_start - subject_start) / 3)\n",
    "phase_map = {0:0, 1:2, 2:1}\n",
    "phase = phase_map[abs(query_start - subject_start) % 3]\n",
    "print('phase', phase)\n",
    "print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "caa1b2b0",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'log_totfigure_df.phase_start == figure_df.phase_stopal'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/frugal/lib/python3.13/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'log_totfigure_df.phase_start == figure_df.phase_stopal'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 46\u001b[0m\n\u001b[1;32m     43\u001b[0m     fig\u001b[38;5;241m.\u001b[39mtight_layout()\n\u001b[1;32m     44\u001b[0m     plt\u001b[38;5;241m.\u001b[39mshow()\n\u001b[0;32m---> 46\u001b[0m \u001b[43mfigure_1\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdataset_df\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     47\u001b[0m \u001b[38;5;66;03m# Worth noting that a sequence being registered as a conflict implies that it does not have a match, as \u001b[39;00m\n\u001b[1;32m     48\u001b[0m \u001b[38;5;66;03m# the top hit selection prioritizes overlaps which are in-frame. \u001b[39;00m\n",
      "Cell \u001b[0;32mIn[10], line 15\u001b[0m, in \u001b[0;36mfigure_1\u001b[0;34m(dataset_df)\u001b[0m\n\u001b[1;32m     13\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfraction\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m dataset_df\u001b[38;5;241m.\u001b[39mgroupby(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcategory\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m df : df\u001b[38;5;241m.\u001b[39mtop_hit_unsupported\u001b[38;5;241m.\u001b[39msum() \u001b[38;5;241m/\u001b[39m \u001b[38;5;28mlen\u001b[39m(df), include_groups\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m     14\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124munsupported\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlog_total\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m*\u001b[39m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfraction\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m---> 15\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mconfirmed\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43max_a_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mlog_totfigure_df.phase_start == figure_df.phase_stopal\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m*\u001b[39m (\u001b[38;5;241m1\u001b[39m \u001b[38;5;241m-\u001b[39m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfraction\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m     17\u001b[0m positions \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marange(dataset_df\u001b[38;5;241m.\u001b[39mcategory\u001b[38;5;241m.\u001b[39mnunique())\n\u001b[1;32m     18\u001b[0m ax_a\u001b[38;5;241m.\u001b[39mbar(positions, ax_a_df\u001b[38;5;241m.\u001b[39munsupported, color\u001b[38;5;241m=\u001b[39max_a_df\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39mmap(palette), edgecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mblack\u001b[39m\u001b[38;5;124m'\u001b[39m, hatch\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m///\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/frugal/lib/python3.13/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/envs/frugal/lib/python3.13/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'log_totfigure_df.phase_start == figure_df.phase_stopal'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0EAAAGvCAYAAACdLN4CAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAHstJREFUeJzt3W9snWX9+PHPtrLOnranjiJkaSdDB67rlARwksz/MQxBnUGpA6YheyAuaBo16oJ1zkVmiDoxaIAHOpARGJ1MgpOYGQrxX0rEEFhBGE47N7fplrWd1WrZ/X3w+62h7JzRe7SnW6/XKzkPzs3V06tX1vvDe/2zaVmWZQEAAJCI6ZO9AQAAgEoSQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJOekI+sc//hFvfvObo6urq+yabdu2xaJFi6JQKMSCBQvi4YcfPtl3BwAnZC4BMFYnFUG/+c1v4tJLL40XX3yx7JoXXnghrrrqqli3bl309fXF2rVr4+qrr449e/ac9GYBoBRzCYA8ckfQXXfdFddcc01885vffNV173znO2PZsmVRVVUVV199dbz73e+OO++886Q3CwCvZC4BkFdV3je47LLL4tprr42qqqr4xCc+UXbdjh07YtGiRaOutbS0xFNPPVVy/dDQUAwNDY08P3r0aBw6dCjOPPPMmDZtWt5tAnCSsiyLgYGBmDNnTkyffur/6OhEzaUIswngVDHesyl3BJ1zzjljWjcwMBCFQmHUtZqamjhy5EjJ9evXr4+1a9fm3Q4AE2T37t3R1NQ02dt4VRM1lyLMJoBTzXjNptwRNFaFQiEGBwdHXRscHIy6urqS61evXh2f//znR5739fXF3LlzY/fu3VFfXz9R2wTgFfr7+6O5ubns/fp0lXcuRZhNAKeK8Z5NExZBra2t8eSTT4661tPTExdffHHJ9dXV1VFdXX3c9fr6eoMGYBJMtW/3yjuXIswmgFPNeM2mCftm7xUrVkRXV1ds3rw5hoeHY/PmzdHV1RUrVqyYqHcJAGWZSwAcM64RVFtbG5s2bYqIiLe85S2xdevWuPnmm+P1r399fOMb34gtW7bE+eefP57vEgDKMpcAKGValmXZZG+ilP7+/igWi9HX1+dbDgAqyP23PGcDMDnG+/576v/uUwAAgHEkggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKbkj6MCBA7Fs2bJoaGiIxsbGaG9vj+Hh4ZJrb7311pg3b17U19fHW9/61tiyZctr3jAAvJy5BEBeuSOora0tamtrY+/evdHd3R3bt2+PDRs2HLfuF7/4Rdx8883xyCOPRH9/f6xZsyauvvrq+Mtf/jIe+waAiDCXAMgvVwTt3Lkzurq64pZbbomampo477zzoqOjI2677bbj1j777LORZVkcPXo0siyLGTNmxMyZM6OqqmrcNg9A2swlAE5Grjv/jh07Yvbs2TFnzpyRay0tLdHb2xuHDx+OhoaGkevLly+PH//4x9HS0hIzZsyIadOmxT333BNNTU0lX3toaCiGhoZGnvf39+f8UABIzUTOpQizCWCqyvWVoIGBgSgUCqOu1dTURETEkSNHRl3/73//GxdeeGF0d3fH4OBg3HnnnbFy5cp4+umnS772+vXro1gsjjyam5vzbA2ABE3kXIowmwCmqlwRVCgUYnBwcNS1Y8/r6upGXb/xxhtj4cKFcckll8TMmTPj+uuvj0svvTQ2btxY8rVXr14dfX19I4/du3fn2RoACZrIuRRhNgFMVbkiqLW1NQ4ePBj79+8fudbT0xNNTU1RLBZHre3t7R31LQQREWeccUbMnDmz5GtXV1dHfX39qAcAnMhEzqUIswlgqsoVQfPnz48lS5ZEe3t7DAwMxK5du2LdunWxcuXK49Z++MMfjttuuy2efPLJOHr0aHR2dsajjz4abW1t47Z5ANJmLgFwMnL/SpzOzs648cYbY968eTF9+vT45Cc/GR0dHRERUVtbG3fccUdce+21sWbNmpgxY0ZcddVVcejQoZg/f35s3bo1LrzwwvH+GABImLkEQF7TsizLJnsTpfT390exWIy+vj7ffgBQQe6/5TkbgMkx3vff3P9YKgAAwOlMBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUnJH0IEDB2LZsmXR0NAQjY2N0d7eHsPDwyXXPvbYY7F48eKora2N5ubmWL9+/WveMAC8nLkEQF65I6itrS1qa2tj79690d3dHdu3b48NGzYct+65556LD37wg7Fq1aoYGBiIn//85/Gd73wnOjs7x2XjABBhLgGQ37Qsy7KxLt65c2fMnz8/9uzZE3PmzImIiPvvvz++9KUvxV//+tdRaz/72c/GoUOHYtOmTSPXnn/++aivr49zzjnnVd9Xf39/FIvF6Ovri/r6+rFuEYDX6HS6/1ZyLkWcXmcDMJWM9/0311eCduzYEbNnzx4ZNBERLS0t0dvbG4cPHx61tru7O84999xYvnx5NDY2xoIFC6Krq6vsoBkaGor+/v5RDwA4kYmcSxFmE8BUlSuCBgYGolAojLpWU1MTERFHjhwZdf3QoUPx/e9/P6677rrYt29f3HHHHfHFL36x7LcdrF+/PorF4sijubk5z9YASNBEzqUIswlgqsoVQYVCIQYHB0ddO/a8rq5u1PXq6ur4yEc+EldccUVUVVXFu971rlixYkVs3ry55GuvXr06+vr6Rh67d+/OszUAEjSRcynCbAKYqqryLG5tbY2DBw/G/v374+yzz46IiJ6enmhqaopisThqbUtLSwwNDY269tJLL0W5H0Gqrq6O6urqPNsBIHETOZcizCaAqSrXV4Lmz58fS5Ysifb29hgYGIhdu3bFunXrYuXKlcetveGGG2Lr1q1xzz33RJZl8fjjj8emTZtixYoV47Z5ANJmLgFwMnL/iuzOzs4YHh6OefPmxeLFi2Pp0qXR0dERERG1tbUjv3Xnfe97Xzz00ENx6623RrFYjOuvvz6+/e1vx4c//OHx/QgASJq5BEBeuX5FdiX5NaQAk8P9tzxnAzA5JvVXZAMAAJzuRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFJEEAAAkBQRBAAAJCV3BB04cCCWLVsWDQ0N0djYGO3t7TE8PHzCt3nmmWeipqYmurq6TnafAFCSuQRAXrkjqK2tLWpra2Pv3r3R3d0d27dvjw0bNpRdPzg4GMuXL49///vfr2mjAFCKuQRAXrkiaOfOndHV1RW33HJL1NTUxHnnnRcdHR1x2223lX2bVatWxUc/+tHXvFEAeCVzCYCTkSuCduzYEbNnz445c+aMXGtpaYne3t44fPjwcevvvvvu2LlzZ6xZs+ZVX3toaCj6+/tHPQDgRCZyLkWYTQBTVa4IGhgYiEKhMOpaTU1NREQcOXJk1PXnnnsubrrpprj33ntjxowZr/ra69evj2KxOPJobm7OszUAEjSRcynCbAKYqnJFUKFQiMHBwVHXjj2vq6sbufaf//wn2tra4nvf+17MnTt3TK+9evXq6OvrG3ns3r07z9YASNBEzqUIswlgqsoVQa2trXHw4MHYv3//yLWenp5oamqKYrE4cu2JJ56I559/PlauXBkNDQ3R0NAQERFXXnllrFq1quRrV1dXR319/agHAJzIRM6lCLMJYKqalmVZlucN3vnOd0ZTU1Pceeed8c9//jM+9KEPxcc+9rH4+te/fuJ3NG1aPProo/Ge97xnTO+nv78/isVi9PX1GToAFXS63X8rNZciTr+zAZgqxvv+m/tXZHd2dsbw8HDMmzcvFi9eHEuXLo2Ojo6IiKitrY1Nmza95k0BwFiZSwDklfsrQZXib9sAJof7b3nOBmByTPpXggAAAE5nIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkpI7gg4cOBDLli2LhoaGaGxsjPb29hgeHi659vbbb48LLrgg6urq4vzzz48f/vCHr3nDAPBy5hIAeeWOoLa2tqitrY29e/dGd3d3bN++PTZs2HDcuq1bt8bq1avjrrvuiv7+/rjrrrvipptuii1btozLxgEgwlwCIL9cEbRz587o6uqKW265JWpqauK8886Ljo6OuO22245bu3fv3vjKV74S73jHO2LatGlx6aWXxnvf+954/PHHx23zAKTNXALgZFTlWbxjx46YPXt2zJkzZ+RaS0tL9Pb2xuHDh6OhoWHk+qpVq0a97YEDB+Lxxx+P7373uyVfe2hoKIaGhkae9/f359kaAAmayLkUYTYBTFW5vhI0MDAQhUJh1LWampqIiDhy5EjZt9u3b19cfvnlcdFFF8U111xTcs369eujWCyOPJqbm/NsDYAETeRcijCbAKaqXBFUKBRicHBw1LVjz+vq6kq+ze9///u45JJL4oILLoiHHnooqqpKf/Fp9erV0dfXN/LYvXt3nq0BkKCJnEsRZhPAVJUrglpbW+PgwYOxf//+kWs9PT3R1NQUxWLxuPU/+tGP4v3vf3+0t7fHvffeG9XV1WVfu7q6Ourr60c9AOBEJnIuRZhNAFNVrgiaP39+LFmyJNrb22NgYCB27doV69ati5UrVx63dsuWLfGZz3wmfvrTn8YXvvCFcdswABxjLgFwMnL/iuzOzs4YHh6OefPmxeLFi2Pp0qXR0dERERG1tbWxadOmiIhYu3ZtDA8Px1VXXRW1tbUjjxtuuGF8PwIAkmYuAZDXtCzLssneRCn9/f1RLBajr6/Ptx8AVJD7b3nOBmByjPf9N/dXggAAAE5nIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkiKCAACApIggAAAgKSIIAABIiggCAACSIoIAAICkiCAAACApIggAAEiKCAIAAJIiggAAgKSIIAAAICkiCAAASIoIAgAAkpI7gg4cOBDLli2LhoaGaGxsjPb29hgeHi65dtu2bbFo0aIoFAqxYMGCePjhh1/zhgHg5cwlAPLKHUFtbW1RW1sbe/fuje7u7ti+fXts2LDhuHUvvPBCXHXVVbFu3bro6+uLtWvXxtVXXx179uwZl40DQIS5BEB+07Isy8a6eOfOnTF//vzYs2dPzJkzJyIi7r///vjSl74Uf/3rX0et/epXvxrd3d3xy1/+cuTa5ZdfHm9/+9tj7dq1r/q++vv7o1gsRl9fX9TX1491iwC8RqfT/beScyni9DobgKlkvO+/VXkW79ixI2bPnj0yaCIiWlpaore3Nw4fPhwNDQ2j1i5atGjU27e0tMRTTz1V8rWHhoZiaGho5HlfX19E/L8PGIDKOXbfzfF3ZJNmIudShNkEcKoY79mUK4IGBgaiUCiMulZTUxMREUeOHBk1bMqtPXLkSMnXXr9+fcm/iWtubs6zRQDGycGDB6NYLE72Nk5oIudShNkEcKoZr9mUK4IKhUIMDg6OunbseV1d3ZjWvnLdMatXr47Pf/7zI88PHz4cb3zjG6O3t/eUH8KV1t/fH83NzbF7927fjvEyzqU051Kesymtr68v5s6dG7Nnz57srbyqiZxLEWbTWPlcKs/ZlOZcynM2pY33bMoVQa2trXHw4MHYv39/nH322RER0dPTE01NTccNg9bW1njyySdHXevp6YmLL7645GtXV1dHdXX1cdeLxaI/AGXU19c7mxKcS2nOpTxnU9r06af+v6IwkXMpwmzKy+dSec6mNOdSnrMpbbxmU65XmT9/fixZsiTa29tjYGAgdu3aFevWrYuVK1cet3bFihXR1dUVmzdvjuHh4di8eXN0dXXFihUrxmXjAGAuAXAycqdUZ2dnDA8Px7x582Lx4sWxdOnS6OjoiIiI2tra2LRpU0REvOUtb4mtW7fGzTffHK9//evjG9/4RmzZsiXOP//88f0IAEiauQRAXrm+HS4i4uyzz44HHnig5H975Q+XXnbZZXHZZZed1Maqq6tjzZo1Jb8NIXXOpjTnUppzKc/ZlHa6nUul5lLE6Xc2leJcynM2pTmX8pxNaeN9Lrn+nSAAAIDT3an/U68AAADjSAQBAABJEUEAAEBSJjWCDhw4EMuWLYuGhoZobGyM9vb2GB4eLrl227ZtsWjRoigUCrFgwYJ4+OGHK7zbyslzLrfffntccMEFUVdXF+eff3788Ic/rPBuKyvP2RzzzDPPRE1NTXR1dVVmk5Mgz7k89thjsXjx4qitrY3m5uZYv359hXdbWXnO5tZbb4158+ZFfX19vPWtb40tW7ZUeLeV949//CPe/OY3n/DzI6X7b4TZVI7ZVJ7ZVJrZVJ7ZdGIVmU3ZJHrPe96TXXvttdm//vWv7MUXX8wWLlyY3XLLLcete/7557NZs2ZlDz74YPa///0vu//++7PXve512d/+9rdJ2PXEG+u5PPjgg1lDQ0P2u9/9Ljt69Gj229/+NmtoaMg6OzsnYdeVMdazOeZf//pX1tramkVE9uijj1ZuoxU21nN59tlns5qammzjxo3Z0aNHs6eeeio788wzswceeGASdl0ZYz2bbdu2ZW94wxuy5557LsuyLOvs7MymT5+e7dq1q8I7rpxf//rX2Zve9KYTfn6kdv/NMrOpHLOpPLOpNLOpPLOpvErNpkmLoBdeeCGLiGzPnj0j1+67775s7ty5x6296aabsg984AOjri1dujT72te+NuH7rLQ85/KDH/wg+9a3vjXq2kc/+tHsc5/73ITvczLkOZtjPvWpT2UdHR1TetDkOZcbb7wxu+aaa0Zd+9Of/pT9/e9/n/B9ToY8Z/Od73wnO+uss7Kenp7s6NGj2YMPPpjNmjUr2717dyW3XDEbN27M5s6dm913330n/PxI6f6bZWZTOWZTeWZTaWZTeWZTeZWcTZP27XA7duyI2bNnx5w5c0autbS0RG9vbxw+fPi4tYsWLRp1raWlJZ566qlKbLWi8pzLqlWr4stf/vLI8wMHDsTjjz8eF110UaW2W1F5ziYi4u67746dO3fGmjVrKrjLystzLt3d3XHuuefG8uXLo7GxMRYsWBBdXV1xzjnnVHjXlZHnbJYvXx5nn312tLS0xBlnnBEf//jHY+PGjdHU1FThXVfGZZddFi+++GK0tbWdcF1K998Is6kcs6k8s6k0s6k8s6m8Ss6mSYuggYGBKBQKo67V1NRExPH/uF25ta9cNxXkOZeX27dvX1x++eVx0UUXxTXXXDOhe5wsec7mueeei5tuuinuvffemDFjRsX2OBnynMuhQ4fi+9//flx33XWxb9++uOOOO+KLX/xidHZ2Vmy/lZTnbP773//GhRdeGN3d3TE4OBh33nlnrFy5Mp5++umK7beSzjnnnKiqevV/Lzul+2+E2VSO2VSe2VSa2VSe2VReJWfTpEVQoVCIwcHBUdeOPa+rqxvT2leumwrynMsxv//97+OSSy6JCy64IB566KEx/eE5HY31bP7zn/9EW1tbfO9734u5c+dWdI+TIc+fmerq6vjIRz4SV1xxRVRVVcW73vWuWLFiRWzevLli+62kPGdz4403xsKFC+OSSy6JmTNnxvXXXx+XXnppbNy4sVLbPSWldP+NMJvKMZvKM5tKM5vKM5teu/G4/05aBLW2tsbBgwdj//79I9d6enqiqakpisXicWt37Ngx6lpPT0+0trZWZK+VlOdcIiJ+9KMfxfvf//5ob2+Pe++9N6qrqyu53Yoa69k88cQT8fzzz8fKlSujoaEhGhoaIiLiyiuvjFWrVlV62xMuz5+ZlpaWGBoaGnXtpZdeiizLKrLXSstzNr29vcedzRlnnBEzZ86syF5PVSndfyPMpnLMpvLMptLMpvLMptduXO6/J/+jS6/dkiVLsk984hNZf39/9uc//zlbuHBhtmbNmuPWPfvss9msWbOy+++/f+Q3QMyaNSv705/+VPlNV8BYz6WzszObOXNm9sgjj1R+k5NkrGfzSjGFf/g0y8Z+Lr/61a+yqqqq7Cc/+Ul29OjR7LHHHstqa2uzn/3sZ5XfdIWM9Wy++tWvZo2Njdkf/vCH7KWXXsoeeOCBbNasWdkf//jHiu+50k70+ZHa/TfLzKZyzKbyzKbSzKbyzKZXN9GzaVIjaN++fdnHPvax7Mwzz8zOOuus7Atf+EI2PDycZVmWFQqF7J577hlZ+8gjj2Rve9vbstra2mzhwoXZz3/+88na9oQb67ksWrQomz59elYoFEY9Pv3pT0/m9idUnj8zLzfVB02ec9m2bVt28cUXZ3V1ddl5552X3X777ZO17YoY69n873//y9asWZOde+65WX19fXbRRRcl8z9xr/z8SPn+m2VmUzlmU3lmU2lmU3lm06ub6Nk07f+/EwAAgCRM2s8EAQAATAYRBAAAJEUEAQAASRFBAABAUkQQAACQFBEEAAAkRQQBAABJEUEAAEBSRBAAAJAUEQQAACRFBAEAAEkRQQAAQFL+D+GjA60bfSLPAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1000x500 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Want a figure summarizing Prodigal results. Probably should have separate figures for matches and everything else, \n",
    "# and need a way to solve the fact that there are many more matches than others. Possibly add length distributions to this figure. \n",
    "\n",
    "def figure_1(dataset_df:pd.DataFrame):\n",
    "\n",
    "    fig, (ax_a, ax_b) = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "\n",
    "    palette = {'match':'darkseagreen', 'pseudogene':'lightgray', 'intergenic':'steelblue', 'conflict':'indianred'}\n",
    "\n",
    "    ax_a_df = pd.DataFrame(index=dataset_df.category.unique())\n",
    "    ax_a_df['total'] = dataset_df.category.value_counts()\n",
    "    ax_a_df['log_total'] = np.log10(ax_a_df['total'])\n",
    "    ax_a_df['fraction'] = dataset_df.groupby('category').apply(lambda df : df.top_hit_unsupported.sum() / len(df), include_groups=False)\n",
    "    ax_a_df['unsupported'] = ax_a_df['log_total'] * ax_a_df['fraction']\n",
    "    ax_a_df['confirmed'] = ax_a_df['log_totfigure_df.phase_start == figure_df.phase_stopal'] * (1 - ax_a_df['fraction'])\n",
    "\n",
    "    positions = np.arange(dataset_df.category.nunique())\n",
    "    ax_a.bar(positions, ax_a_df.unsupported, color=ax_a_df.index.map(palette), edgecolor='black', hatch='///')\n",
    "    ax_a.bar(positions, ax_a_df.confirmed, bottom=ax_a_df.unsupported, color=ax_a_df.index.map(palette), edgecolor='black')\n",
    "\n",
    "    ax_a.set_xticks(positions, labels=ax_a_df.index)\n",
    "    ax_a.set_yticks(np.arange(5), labels=[f'$10^{i}$' for i in range(5)])\n",
    "\n",
    "    for x, total, log_total in zip(positions, ax_a_df.total, ax_a_df.log_total):\n",
    "        text = f'{100 * total / ax_a_df.total.sum():.2f}%'\n",
    "        ax_a.text(x, log_total, text, ha='center', fontsize='small', va='bottom')\n",
    "\n",
    "    ax_b_df = dataset_df.copy()\n",
    "\n",
    "    sns.kdeplot(data=ax_b_df, ax=ax_b, x='length', hue='category', palette=palette, common_norm=False)\n",
    "    ax_b.set_ylabel('density')\n",
    "    ax_b.set_xlim(xmin=0, xmax=800)\n",
    "\n",
    "    text = ''\n",
    "    for category, df in ax_b_df.groupby('category'):\n",
    "        total = len(df)\n",
    "        n_smorfs = (df.length < 100).sum()\n",
    "        text += '$p_{' + category + '<100}$ = ' + f'{get_percent(n_smorfs, total)} ($n={n_smorfs}$)\\n'\n",
    "    ax_b.text(0.3, 0.75, text, transform=ax_b.transAxes, ha='left')\n",
    "    sns.move_legend(ax_b, loc='lower right')\n",
    "    ax_b.axvline(x=100, ls='--', lw=0.5, color='black')\n",
    "\n",
    "    fig.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "figure_1(dataset_df)\n",
    "# Worth noting that a sequence being registered as a conflict implies that it does not have a match, as \n",
    "# the top hit selection prioritizes overlaps which are in-frame. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "967ac4d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "matches 96.51% (n=49422)\n",
      "unsupported_matches 12.31% (n=6306)\n",
      "exact_matches 85.68% (n=43875)\n",
      "extended 6.37% (n=3264)\n",
      "truncated 4.42% (n=2262)\n",
      "pseudogenes 1.06% (n=543)\n"
     ]
    }
   ],
   "source": [
    "def si_table_1(dataset_df:pd.DataFrame):\n",
    "\n",
    "    table_df = pd.DataFrame(index=dataset_df.genome_id.unique())\n",
    "    table_df['phylum'] = dataset_df.groupby('genome_id').phylum.first()\n",
    "    table_df['gc_percent'] = dataset_df.groupby('genome_id').gc_percent.first()\n",
    "    table_df['genome_size'] = dataset_df.groupby('genome_id').apply(len, include_groups=False)\n",
    "    table_df['n_matches'] = dataset_df.groupby('genome_id').match.sum()\n",
    "    table_df['n_unsupported_matches'] = dataset_df.groupby('genome_id').apply(lambda df : ((df.top_hit_unsupported) & df.match).sum(), include_groups=False)\n",
    "    table_df['n_exact_matches'] = dataset_df.groupby('genome_id').exact_match.sum()\n",
    "    table_df['n_extended'] = dataset_df.groupby('genome_id').extended.sum()\n",
    "    table_df['n_truncated'] = dataset_df.groupby('genome_id').truncated.sum()\n",
    "    table_df['n_pseudogenes'] = dataset_df.groupby('genome_id').pseudogene.sum()\n",
    "\n",
    "    totals = table_df.drop(columns=['gc_percent', 'phylum']).sum(axis=0)\n",
    "    for col in [col for col in totals.index if col.startswith('n_')]:\n",
    "        print(col.replace('n_', '', count=1), get_percent(totals.loc[col], totals.genome_size), f'(n={totals.loc[col]})')\n",
    "\n",
    "    for col in [col for col in table_df.columns if col.startswith('n_')]:\n",
    "        table_df[col.replace('n_', '', count=1)] = [f'{n} ({get_percent(n, total)})' for n, total in zip(table_df[col], table_df.genome_size)]\n",
    "\n",
    "    table_df = table_df.sort_values('gc_percent')\n",
    "    table_df.index = table_df.index.map(dataset_df.groupby('genome_id').species.first())\n",
    "\n",
    "    table_df.drop(columns=[col for col in table_df.columns if col.startswith('n_')])\n",
    "\n",
    "si_table_1(dataset_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dd1eabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def figure_2(results_df:pd.DataFrame):\n",
    "\n",
    "    fig, (ax_a, ax_b) = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "\n",
    "    figure_df = results_df.copy()\n",
    "    figure_df['category'] = np.select([results_df.match, results_df.pseudogene, results_df.intergenic, results_df.conflict], ['match', 'pseudogene', 'intergenic', 'conflict'], default='none')\n",
    "    categories = ['match', 'conflict', 'intergenic', 'pseudogene']\n",
    "    figure_df['category'] = pd.Categorical(figure_df.category, categories, ordered=True)\n",
    "\n",
    "    width = 0.45\n",
    "    offsets = {0: -width/2, 1:width/2}\n",
    "    palette = {0:'indianred', 1:'darkseagreen'}\n",
    "\n",
    "    for model_label, df in figure_df.groupby('model_label'):\n",
    "        ax_df = pd.DataFrame(index=categories)\n",
    "        ax_df['total'] = df.category.value_counts()\n",
    "        ax_df['log_total'] = np.log10(ax_df['total'])\n",
    "        ax_df['fraction'] = df.groupby('category', observed=True).apply(lambda df : df.top_hit_unsupported.sum() / len(df), include_groups=False)\n",
    "        ax_df['unsupported'] = ax_df['log_total'] * ax_df['fraction']\n",
    "        ax_df['confirmed'] = ax_df['log_total'] * (1 - ax_df['fraction'])\n",
    "\n",
    "        positions = np.arange(df.category.nunique())\n",
    "        ax_a.bar(positions + offsets[model_label], ax_df.unsupported, color=palette[model_label], edgecolor='black', hatch='///', width=width)\n",
    "        ax_a.bar(positions + offsets[model_label], ax_df.confirmed, bottom=ax_df.unsupported, color=palette[model_label], edgecolor='black', width=width)\n",
    "\n",
    "        for x, total, log_total in zip(positions + offsets[model_label], ax_df.total, ax_df.log_total):\n",
    "            text = f'{100 * total / ax_df.total.sum():.2f}%'\n",
    "            text = total\n",
    "            ax_a.text(x, log_total, text, ha='center', fontsize='small', va='bottom')\n",
    "\n",
    "    ax_a.set_xticks(positions, labels=ax_df.index)\n",
    "    ax_a.set_yticks(np.arange(5), labels=[f'$10^{i}$' for i in range(5)])\n",
    "\n",
    "\n",
    "    fig.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "figure_2(results_df)\n",
    "\n",
    "# I think should probably combine these bar plots into a single chart. Although there are so many more real than spurious sequences, this might look a bit weird. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "frugal",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
