{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.tools import MMseqs, NCBIDatasets, UniRef\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from src.files import XMLFile, InterProScanFile, FASTAFile, GBFFFile\n",
    "from utils import * \n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split, GroupShuffleSplit\n",
    "from src.classifier import Classifier\n",
    "from src.sampler import Sampler\n",
    "from src.genome import ReferenceGenome\n",
    "from src.dataset import Dataset\n",
    "import re\n",
    "from src.build import * \n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phylum count: 56\n",
      "Class count: 123\n",
      "Order count: 299\n",
      "Genus count: 3983\n"
     ]
    }
   ],
   "source": [
    "ncbi_genome_metadata_df = load_ncbi_genome_metadata()\n",
    "\n",
    "for level in ['phylum', 'class', 'order', 'genus']:\n",
    "    print(f'{level.capitalize()} count:', ncbi_genome_metadata_df[f'{level}'].nunique())\n",
    "    # print(f'Species with no {level}:', (genome_metadata_df[f'{level}_taxid'] == 'none').sum())\n",
    "\n",
    "# fig, (ax_a, ax_b) = plt.subplots(ncols=2, figsize=(15, 5), width_ratios=[0.3, 0.7])\n",
    "\n",
    "# counts, bins, _ = ax_a.hist(ncbi_genome_metadata_df.gc_percent, bins=25, width=2, edgecolor='black', color='lightgray')\n",
    "# ax_a.set_ylabel('density')\n",
    "# ax_a.set_xlabel('GC percent')\n",
    "\n",
    "# level = 'phylum'\n",
    "# # x = np.arange(genome_metadata_df[f'{level}_taxid'].nunique())\n",
    "# y = ncbi_genome_metadata_df.groupby(level).gc_percent.mean() # .sort_values(ascending=True)\n",
    "# x = ncbi_genome_metadata_df.groupby(level).apply(len, include_groups=False) # .sort_values(ascending=True)\n",
    "# y_err = ncbi_genome_metadata_df.groupby(level).apply(lambda df : df.gc_percent.std() / np.sqrt(len(df)), include_groups=False) # .loc[y.index]\n",
    "\n",
    "# ax_b.errorbar(x, y, yerr=y_err, ls='', lw=0.7, capsize=2, color='black')\n",
    "# ax_b.scatter(x, y, color='black', s=5)\n",
    "# ax_b.set_xscale('log')\n",
    "# ax_b.set_xlabel(f'log({level} size)')\n",
    "# ax_b.set_ylabel('GC percent')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Low-GC content phylum: Campylobacterota\n",
      "Medium-GC content phylum: Planctomycetota\n",
      "High-GC content phylum: Deinococcota\n"
     ]
    }
   ],
   "source": [
    "# How should I select the genomes for model training?\n",
    "\n",
    "# Using the same genomes for validation as in the Prodigal paper\n",
    "# Halobacterium salinarum\n",
    "# Natronomonas pharaonis\n",
    "# Aeropyrum pernix\n",
    "model_organism_genome_ids = ['GCF_000005845.2'] # E. coli K-12\n",
    "model_organism_genome_ids += ['GCF_000009045.1'] # B. subtilis\n",
    "model_organism_genome_ids += ['GCF_000006765.1'] # P. aeruginosa\n",
    "\n",
    "# Start by selecting a few phyla in a reasonable size range with varying GC contents.\n",
    "def sample(genome_metadata_df:pd.DataFrame, min_phylum_size:int=90, max_phylum_size:int=200):\n",
    "\n",
    "    phylum_sizes = genome_metadata_df.groupby('phylum').apply(len, include_groups=False)\n",
    "    phylum_sizes = phylum_sizes[(phylum_sizes > min_phylum_size) & (phylum_sizes < max_phylum_size)]\n",
    "\n",
    "    genome_metadata_df = genome_metadata_df[genome_metadata_df.phylum.isin(phylum_sizes.index)]\n",
    "    \n",
    "    phylum_gc_percent = genome_metadata_df.groupby('phylum').gc_percent.mean() \n",
    "    phylum_gc_percent = phylum_gc_percent.sort_values(ascending=True)\n",
    "    \n",
    "    phyla = dict()\n",
    "    phyla['low_gc_content'] = phylum_gc_percent.index[0]\n",
    "    phyla['med_gc_content'] = phylum_gc_percent.index[len(phylum_gc_percent) // 2]\n",
    "    phyla['high_gc_content'] = phylum_gc_percent.index[-1]\n",
    "\n",
    "    print('Low-GC content phylum:', phyla['low_gc_content'])\n",
    "    print('Medium-GC content phylum:', phyla['med_gc_content'])\n",
    "    print('High-GC content phylum:', phyla['high_gc_content'])\n",
    "\n",
    "    genome_ids = {category:genome_metadata_df.index[genome_metadata_df.phylum == phylum] for category, phylum in phyla.items()}\n",
    "    return genome_ids\n",
    "\n",
    "genome_ids = sample(ncbi_genome_metadata_df)\n",
    "\n",
    "# ncbi = NCBIDatasets(genome_dir='../data/genomes', gbff_dir='../data/proteins/ncbi')\n",
    "# ncbi.run(genome_ids=genome_ids['low_gc_content'], metadata_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_metadata_df = list()\n",
    "for category, genome_ids_ in genome_ids.items():\n",
    "    df = ncbi_genome_metadata_df.loc[genome_ids_, :].copy()\n",
    "    df['gc_percent_category'] = category\n",
    "    # df = df[genome_metadata_df.annotation_pipeline.str.contains('PGAP')] # Only include genomes annotated with PGAP, because the other files are not consistent. \n",
    "    genome_metadata_df.append(df)\n",
    "genome_metadata_df = pd.concat(genome_metadata_df)\n",
    "# genome_metadata_df.to_csv('../data/genome_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from src.tools import download_homologs\n",
    "# for path in glob.glob('../data/proteins/ncbi/*'):\n",
    "#     download_homologs(path, pseudo_only=True, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref: Searching reference for GCF_001753675.2.: 100%|█| 1/1 [00:00<00:00, 12520.3\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_002443295.1.: 100%|█| 1/1 [00:00<00:00, 12945.3\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000949635.1.: 100%|█| 1/1 [00:00<00:00, 10591.6\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000346315.1.: 100%|█| 1/1 [00:00<00:00, 13273.1\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_010367455.1.: 100%|█| 1/1 [00:00<00:00, 12557.8\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_013036045.1.: 100%|█| 1/1 [00:00<00:00, 12633.4\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_025060895.1.: 100%|█| 1/1 [00:00<00:00, 11428.6\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_024129865.1.: 100%|█| 1/1 [00:00<00:00, 9597.95\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_011064615.1.: 100%|█| 1/1 [00:00<00:00, 10180.3\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_009618275.1.: 100%|█| 1/1 [00:00<00:00, 11155.0\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_034097035.1.: 100%|█| 1/1 [00:00<00:00, 11244.7\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_003977685.1.: 100%|█| 1/1 [00:00<00:00, 10727.1\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859735.1.: 100%|█| 1/1 [00:00<00:00, 11335.9\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_023701485.1.: 100%|█| 1/1 [00:00<00:00, 11335.9\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_030012215.1.: 100%|█| 1/1 [00:00<00:00, 11184.8\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_001707835.1.: 100%|█| 1/1 [00:00<00:00, 11037.6\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_014192335.1.: 100%|█| 1/1 [00:00<00:00, 11125.4\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859855.1.: 100%|█| 1/1 [00:00<00:00, 11305.4\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_011064595.1.: 100%|█| 1/1 [00:00<00:00, 7738.57\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_039541335.1.: 100%|█| 1/1 [00:00<00:00, 3057.07\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_002727185.1.: 100%|█| 1/1 [00:00<00:00, 7516.67\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859915.1.: 100%|█| 1/1 [00:00<00:00, 7206.71\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_011065895.1.: 100%|█| 1/1 [00:00<00:00, 7796.10\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_013694095.1.:   0%|      | 0/1 [00:00<?, ?it/s]/home/prichter/miniconda3/envs/tripy/lib/python3.13/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_013694095.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_013694095.1.: 100%|█| 1/1 [00:14<00:00, 14.62s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_024054565.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_024054565.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_024054565.1.: 100%|█| 1/1 [00:21<00:00, 21.57s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_009177065.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_009177065.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_009177065.1.: 100%|█| 1/1 [00:28<00:00, 28.50s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_900113665.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_900113665.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_900113665.1.: 100%|█| 1/1 [00:23<00:00, 23.41s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007860135.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007860135.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007860135.1.: 100%|█| 1/1 [00:38<00:00, 38.95s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_001027925.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.add_homologs: Loaded 13 homologs; 42 pseudogenes present in the genome.\n",
      "ref: Searching reference for GCF_001027925.1.: 100%|█| 1/1 [00:23<00:00, 23.97s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_039545035.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_039545035.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_039545035.1.: 100%|█| 1/1 [00:23<00:00, 23.42s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_013618625.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_013618625.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_013618625.1.: 100%|█| 1/1 [00:28<00:00, 28.49s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_008629675.1.:   0%|      | 0/1 [00:00<?, ?it/s]/home/prichter/miniconda3/envs/tripy/lib/python3.13/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_008629675.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_008629675.1.: 100%|█| 1/1 [00:37<00:00, 37.82s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m40\u001b[0m\n",
      "    self.weights = \u001b[1;31mtorch.FloatTensor(weights).to(DEVICE) if(weights is None\u001b[0m)\n",
      "                   \u001b[1;31m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35mexpected 'else' after 'if' expression\u001b[0m\n",
      "ref: Searching reference for GCF_022267515.1.: 100%|█| 1/1 [00:00<00:00, 11335.9\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007860095.1.: 100%|█| 1/1 [00:00<00:00, 11586.4\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_012958305.1.: 100%|█| 1/1 [00:00<00:00, 9822.73\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_900182915.1.: 100%|█| 1/1 [00:00<00:00, 10645.4\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m43\u001b[0m\n",
      "    self.\u001b[1;31m\u001b[0m\n",
      "         \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m43\u001b[0m\n",
      "    self.\u001b[1;31m\u001b[0m\n",
      "         \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m43\u001b[0m\n",
      "    self.user_specified_weights = \u001b[1;31m\u001b[0m\n",
      "                                  \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "ref: Searching reference for GCF_007860125.1.: 100%|█| 1/1 [00:00<00:00, 6921.29\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m47\u001b[0m\n",
      "    if \u001b[1;31m\u001b[0m\n",
      "       \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m47\u001b[0m\n",
      "    if \u001b[1;31m\u001b[0m\n",
      "       \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m48\u001b[0m\n",
      "    if \u001b[1;31m\u001b[0m\n",
      "       \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m48\u001b[0m\n",
      "    if \u001b[1;31m\u001b[0m\n",
      "       \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m5\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.cli import ref\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    from src.classifier import Classifier\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/classifier.py\"\u001b[0m, line \u001b[35m48\u001b[0m\n",
      "    if \u001b[1;31m\u001b[0m\n",
      "       \u001b[1;31m^\u001b[0m\n",
      "\u001b[1;35mSyntaxError\u001b[0m: \u001b[35minvalid syntax\u001b[0m\n",
      "ref: Searching reference for GCF_007860175.1.: 100%|█| 1/1 [00:00<00:00, 4788.02\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859765.1.: 100%|█| 1/1 [00:00<00:00, 4888.47\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_011634775.1.: 100%|█| 1/1 [00:00<00:00, 4702.13\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859945.1.: 100%|█| 1/1 [00:00<00:00, 5753.50\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859865.1.: 100%|█| 1/1 [00:00<00:00, 6168.09\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_002197845.1.: 100%|█| 1/1 [00:00<00:00, 6168.09\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_020966755.1.: 100%|█| 1/1 [00:00<00:00, 6423.13\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007860075.1.: 100%|█| 1/1 [00:00<00:00, 11125.4\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_014207595.1.: 100%|█| 1/1 [00:00<00:00, 11781.7\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007859715.1.: 100%|█| 1/1 [00:00<00:00, 10407.7\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_029589435.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.add_homologs: Loaded 24 homologs; 103 pseudogenes present in the genome.\n",
      "ref: Searching reference for GCF_029589435.1.:   0%|      | 0/1 [00:02<?, ?it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \u001b[35m\"/home/prichter/miniconda3/envs/tripy/bin/ref\"\u001b[0m, line \u001b[35m8\u001b[0m, in \u001b[35m<module>\u001b[0m\n",
      "    sys.exit(\u001b[31mref\u001b[0m\u001b[1;31m()\u001b[0m)\n",
      "             \u001b[31m~~~\u001b[0m\u001b[1;31m^^\u001b[0m\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/cli.py\"\u001b[0m, line \u001b[35m119\u001b[0m, in \u001b[35mref\u001b[0m\n",
      "    genome = ReferenceGenome(ref_path, load_homologs=args.load_homologs, homologs_dir=args.homologs_dir)\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/genome/reference.py\"\u001b[0m, line \u001b[35m28\u001b[0m, in \u001b[35m__init__\u001b[0m\n",
      "    \u001b[31mself.load_homologs\u001b[0m\u001b[1;31m(dir_=homologs_dir)\u001b[0m\n",
      "    \u001b[31m~~~~~~~~~~~~~~~~~~\u001b[0m\u001b[1;31m^^^^^^^^^^^^^^^^^^^\u001b[0m\n",
      "  File \u001b[35m\"/home/prichter/Documents/tripy/src/genome/reference.py\"\u001b[0m, line \u001b[35m155\u001b[0m, in \u001b[35mload_homologs\u001b[0m\n",
      "    assert \u001b[1;31mlen(homologs_df) == len(self.df)\u001b[0m, f'ReferenceGenome.load_homologs: Expected len(homologs_df) == len(self.df), but len(homologs_df) is {len(homologs_df)} and len(self.df) is {len(self.df)}.'\n",
      "           \u001b[1;31m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n",
      "\u001b[1;35mAssertionError\u001b[0m: \u001b[35mReferenceGenome.load_homologs: Expected len(homologs_df) == len(self.df), but len(homologs_df) is 6582 and len(self.df) is 6580.\u001b[0m\n",
      "ref: Searching reference for GCF_008374075.1.: 100%|█| 1/1 [00:00<00:00, 12264.0\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_024256385.1.: 100%|█| 1/1 [00:00<00:00, 11618.5\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_017939745.1.: 100%|█| 1/1 [00:00<00:00, 11066.7\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000346295.1.: 100%|█| 1/1 [00:00<00:00, 12018.0\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000346505.1.: 100%|█| 1/1 [00:00<00:00, 7767.23\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000255705.1.: 100%|█| 1/1 [00:00<00:00, 7724.32\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000255655.1.: 100%|█| 1/1 [00:00<00:00, 8405.42\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_039541315.1.: 100%|█| 1/1 [00:00<00:00, 8848.74\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_000153105.1.: 100%|█| 1/1 [00:00<00:00, 8305.55\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_008312935.1.: 100%|█| 1/1 [00:00<00:00, 8322.03\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_008329715.1.: 100%|█| 1/1 [00:00<00:00, 8388.61\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007748155.1.: 100%|█| 1/1 [00:00<00:00, 8439.24\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_008035925.1.: 100%|█| 1/1 [00:00<00:00, 8208.03\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007746795.1.: 100%|█| 1/1 [00:00<00:00, 12826.6\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_008065115.1.: 100%|█| 1/1 [00:00<00:00, 12748.6\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007744515.1.: 100%|█| 1/1 [00:00<00:00, 12865.9\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007748195.1.: 100%|█| 1/1 [00:00<00:00, 11275.0\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007745435.1.: 100%|█| 1/1 [00:00<00:00, 12336.1\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007745175.1.: 100%|█| 1/1 [00:00<00:00, 11335.9\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007754155.1.: 100%|█| 1/1 [00:00<00:00, 10951.1\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007741475.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007741475.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007741475.1.: 100%|█| 1/1 [00:56<00:00, 56.42s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007746075.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.add_homologs: Loaded 12 homologs; 46 pseudogenes present in the genome.\n",
      "ref: Searching reference for GCF_007746075.1.: 100%|█| 1/1 [00:28<00:00, 28.82s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007743815.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007743815.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007743815.1.: 100%|█| 1/1 [00:35<00:00, 35.18s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007744255.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007744255.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007744255.1.: 100%|█| 1/1 [00:41<00:00, 41.12s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007753095.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007753095.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007753095.1.: 100%|█| 1/1 [00:45<00:00, 45.57s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007744675.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.add_homologs: Loaded 6 homologs; 15 pseudogenes present in the genome.\n",
      "ref: Searching reference for GCF_007744675.1.: 100%|█| 1/1 [00:59<00:00, 59.60s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007753265.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007753265.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007753265.1.: 100%|█| 1/1 [00:36<00:00, 36.40s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007754425.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007754425.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007754425.1.: 100%|█| 1/1 [00:54<00:00, 54.03s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007752935.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007752935.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007752935.1.: 100%|█| 1/1 [00:51<00:00, 51.45s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_007752535.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_007752535.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_007752535.1.: 100%|█| 1/1 [01:11<00:00, 71.56s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_041508025.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_041508025.1 was found in ../data/proteins/homologs/\n",
      "ref: Searching reference for GCF_041508025.1.: 100%|█| 1/1 [01:17<00:00, 77.37s/\n",
      "ref: Search complete. Results written to ../data/ref\n",
      "ref: Searching reference for GCF_009002475.1.:   0%|      | 0/1 [00:00<?, ?it/s]\n",
      "ReferenceGenome.load_homologs: No homolog file for genome GCF_009002475.1 was found in ../data/proteins/homologs/\n"
     ]
    }
   ],
   "source": [
    "for genome_id in genome_ids['med_gc_content'][::-1]:\n",
    "    ! ref --input-path ../data/proteins/prodigal/{genome_id}_protein.faa --output-dir ../data/ref --reference-dir ../data/proteins/ncbi/ --prodigal-output --summarize --load-homologs --homologs-dir ../data/proteins/homologs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = load_ref(genome_ids=[genome_id for genome_id in genome_metadata_df.index if (genome_id != 'GCF_000009085.1')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def check_spurious(ref_df:pd.DataFrame, max_score:float=0.5, remove_invalid:bool=True):\n",
    "    scores = get_alignment_scores(ref_df, seq_a_col='top_hit_seq', seq_b_col='query_seq')\n",
    "    # ref_df['alignment_score'] = scores\n",
    "\n",
    "    valid = ref_df.top_hit_feature.isin(GBFFFile.rna_features + GBFFFile.other_features)\n",
    "    valid = valid | ((scores < max_score) & ~np.isnan(scores)) \n",
    "    ref_df['valid'] = valid\n",
    "\n",
    "    print(f'check_spurious: Validated {ref_df.valid.sum()} out of {len(ref_df)} spurious sequences.')\n",
    "\n",
    "    if remove_invalid:\n",
    "        ref_df = ref_df[ref_df.valid].copy()\n",
    "        ref_df = ref_df.drop(columns=['valid'])\n",
    "    return ref_df\n",
    "    \n",
    "\n",
    "def check_real(ref_df:pd.DataFrame, min_score:float=0.9, remove_invalid:bool=True):\n",
    "    scores = get_alignment_scores(ref_df, seq_a_col='top_hit_seq', seq_b_col='query_seq')\n",
    "    # ref_df['alignment_score'] = scores\n",
    "    ref_df['valid'] = (scores > min_score) & ~np.isnan(scores)\n",
    "\n",
    "    print(f'check_real: Validated {ref_df.valid.sum()} out of {len(ref_df)} real sequences.')\n",
    "    \n",
    "    if remove_invalid:\n",
    "        ref_df = ref_df[ref_df.valid].copy()\n",
    "        ref_df = ref_df.drop(columns=['valid'])\n",
    "    return ref_df\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "build_datasets: Loading data from 460 genomes belonging to the phylum Campylobacterota.\n",
      "build_datasets: Loaded 407732 sequences.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "get_alignment_scores: 100%|██████████| 360302/360302 [05:49<00:00, 1030.05it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "check_real: Validated 358960 out of 360302 real sequences.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "get_alignment_scores: 100%|██████████| 1013/1013 [00:00<00:00, 8924.88it/s]\n",
      "/home/prichter/Documents/tripy/notebooks/utils.py:60: UserWarning: get_lengths: Not all gene lengths are divisible by three.\n",
      "  warnings.warn('get_lengths: Not all gene lengths are divisible by three.')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "check_spurious: Validated 1012 out of 1013 spurious sequences.\n",
      "build_datasets: Kept 1012 spurious sequences and 358960 real sequences.\n",
      "Removing 344 sequences exceeding the maximum length of 2000\n",
      "MMseqs.load: Removing 44135 non-cluster representatives.\n",
      "build_datasets: Training dataset contains 252990 sequences, testing dataset contains 62503 sequences.\n",
      "build_datasets: 579 negative instances and 252411 positive instances in the training dataset.\n",
      "build_datasets: 150 negative instances and 62353 positive instances in the testing dataset.\n"
     ]
    }
   ],
   "source": [
    "def build_datasets(genome_metadata_df:pd.DataFrame, phylum:str='Campylobacterota', max_length:int=2000):\n",
    "    # First remove things which do not have definitive labels. \n",
    "    genome_ids = genome_metadata_df[genome_metadata_df.phylum == phylum].index\n",
    "    genome_ids = [genome_id for genome_id in genome_metadata_df.index if (genome_id != 'GCF_000009085.1')] # Remove the non-PGAP genome. \n",
    "\n",
    "    print(f'build_datasets: Loading data from {len(genome_ids)} genomes belonging to the phylum {phylum}.')\n",
    "    df = load_ref(genome_ids=genome_ids)\n",
    "    print(f'build_datasets: Loaded {len(ref_df)} sequences.')\n",
    "\n",
    "    df = pd.concat([check_real(df[df.real].copy()), check_spurious(df[df.spurious].copy())])\n",
    "    print(f'build_datasets: Kept {df.spurious.sum()} spurious sequences and {df.real.sum()} real sequences.')\n",
    "    df['label'] = df.real.astype(int)\n",
    "\n",
    "    lengths = get_lengths(df, top_hit=False)\n",
    "    print(f'Removing {(lengths >= max_length).sum()} sequences exceeding the maximum length of {max_length}')\n",
    "    df = df[lengths < max_length]\n",
    "\n",
    "    df = df.rename(columns={'query_seq':'seq'}) # Need to do this for file writing, etc. to work correctly, \n",
    "    df = df.rename(columns={'top_hit_genome_id':'genome_id'}) # Need to do this for file writing, etc. to work correctly, \n",
    "\n",
    "    mmseqs = MMseqs()\n",
    "    df = mmseqs.cluster(df, job_name=f'{phylum.lower()}_95', sequence_identity=0.95, reps_only=True, overwrite=True)\n",
    "    mmseqs.cleanup()\n",
    "\n",
    "    gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "    train_idxs, test_idxs = list(gss.split(df, groups=df.genome_id))[0]\n",
    "    train_df, test_df = df.iloc[train_idxs], df.iloc[test_idxs]\n",
    "    print(f'build_datasets: Training dataset contains {len(train_df)} sequences, testing dataset contains {len(test_df)} sequences.')\n",
    "    print(f'build_datasets: {(train_df.label == 0).sum()} negative instances and {(train_df.label == 1).sum()} positive instances in the training dataset.')\n",
    "    print(f'build_datasets: {(test_df.label == 0).sum()} negative instances and {(test_df.label == 1).sum()} positive instances in the testing dataset.')\n",
    "    train_df.to_csv(f'../data/{phylum.lower()}_dataset_train.csv')\n",
    "    test_df.to_csv(f'../data/{phylum.lower()}_dataset_test.csv')\n",
    "    return train_df, test_df \n",
    "    \n",
    "train_df, test_df = build_datasets(genome_metadata_df)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['11', 'none'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.top_hit_translation_table.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There still seems to be something wrong with the training data, perhaps I am still misclassifying the spurious sequences?\n",
    "# It is possible my \"in-frame\" check is wrong, it's possible I should only see if either edge is a multiple of 3 base pairs away, \n",
    "# as opposed to both.\n",
    "\n",
    "# It seems as though 684 of the sequences flagged as spurious have an overlap length of 0, so maybe I should be counting those as intergenic?\n",
    "# Ok, so I think because of how frequent gene overlap is, I should be allowing some (considering genes with less than a certain amount of overlap to be intergenic)\n",
    "# I should also be very conservative, i.e. allow a comparatively large amount of overlap. \n",
    "# https://pmc.ncbi.nlm.nih.gov/articles/PMC525685/ \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/prichter/Documents/tripy/notebooks/utils.py:60: UserWarning: get_lengths: Not all gene lengths are divisible by three.\n",
      "  warnings.warn('get_lengths: Not all gene lengths are divisible by three.')\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'top_hit_overlap_length'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/tripy/lib/python3.13/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'top_hit_overlap_length'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 28\u001b[0m\n\u001b[1;32m     24\u001b[0m     plt\u001b[38;5;241m.\u001b[39mshow()\n\u001b[1;32m     25\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ax_a_df\n\u001b[0;32m---> 28\u001b[0m ax_a_df \u001b[38;5;241m=\u001b[39m \u001b[43mfigure\u001b[49m\u001b[43m(\u001b[49m\u001b[43mref_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmin_overlap_fraction\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[12], line 14\u001b[0m, in \u001b[0;36mfigure\u001b[0;34m(ref_df, path, min_overlap_fraction)\u001b[0m\n\u001b[1;32m     12\u001b[0m ax_a_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(index\u001b[38;5;241m=\u001b[39mref_df\u001b[38;5;241m.\u001b[39mindex)\n\u001b[1;32m     13\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlength\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m get_lengths(ref_df, top_hit\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, units\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnt\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 14\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124moverlap_length\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mref_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtop_hit_overlap_length\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     15\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124moverlap_fraction\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m ax_a_df\u001b[38;5;241m.\u001b[39moverlap_length \u001b[38;5;241m/\u001b[39m ax_a_df\u001b[38;5;241m.\u001b[39mlength \u001b[38;5;66;03m# This is relative to the query sequence. \u001b[39;00m\n\u001b[1;32m     16\u001b[0m ax_a_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlabel\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mselect([ref_df\u001b[38;5;241m.\u001b[39mreal, ref_df\u001b[38;5;241m.\u001b[39mspurious], [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreal\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspurious\u001b[39m\u001b[38;5;124m'\u001b[39m], default\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/tripy/lib/python3.13/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/envs/tripy/lib/python3.13/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'top_hit_overlap_length'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA90AAAPdCAYAAACXzguGAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAKwBJREFUeJzt3X+s1nd9//8nLeU05xzOOSLGhpzDSiOtnB5ck9oxEnQ/zGKrTjEoSJUtDX/MkW45UeNGkDEkK0uzDms60/YPA04ay2BjTVubhaWnTUzMMXNpFKwtXRUCAxyEcw45evSU6/vPtyefI1S54DwOYG+35P3H9crrvM/zSl4eufe6zrlmNBqNRgEAAABT7prLPQAAAAD8phLdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAi56Oj+yU9+Uu94xztqYGDgDfc8/fTTtXjx4mpra6tFixbVk08+ebHfDgAAAK46FxXd3/rWt2rp0qX1yiuvvOGel19+uVasWFFbtmypoaGh2rx5c61cubKOHDly0cMCAADA1aTp6N6xY0fdfffd9Xd/93e/dt973vOeWr58ec2cObNWrlxZv/d7v1ePPvroRQ8LAAAAV5OZzX7B+9///vrkJz9ZM2fOrE984hNvuG///v21ePHiSWu9vb31wgsvnHf/2NhYjY2NTTw+e/ZsnTp1qt761rfWjBkzmh0TAAAAmtZoNGpkZKTmzZtX11xz6X8GrenovuGGGy5o38jISLW1tU1aa21trTNnzpx3/9atW2vz5s3NjgMAAABT7vDhw9Xd3X3J92k6ui9UW1tbjY6OTlobHR2t2bNnn3f/+vXr6zOf+czE46GhoZo/f34dPny4Ojo6UmMCAADAhOHh4erp6XnDdm1WLLr7+vrqu9/97qS1AwcO1Lvf/e7z7m9paamWlpZz1js6OkQ3AAAA02qqfs059jnda9asqYGBgdq1a1eNj4/Xrl27amBgoNasWZP6lgAAAHBFmdLobm9vr507d1ZV1Tvf+c7au3dv3XffffWWt7ylvvjFL9aePXvq5ptvnspvCQAAAFesGY1Go3G5hzif4eHh6uzsrKGhIW8vBwAAYFpMdYvG3l4OAAAAb3aiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACENB3dJ06cqOXLl1dXV1fNnTu3+vv7a3x8/Lx7H3zwwVqwYEF1dHTUu971rtqzZ88lDwwAAABXi6aje9WqVdXe3l5Hjx6twcHB2rdvX23btu2cfd/85jfrvvvuq2eeeaaGh4dr06ZNtXLlyvrRj340FXMDAADAFa+p6D548GANDAzU/fffX62trXXTTTfVxo0b66GHHjpn7w9+8INqNBp19uzZajQade2119asWbNq5syZ57332NhYDQ8PT7oAAADgatZUdO/fv7/mzJlT8+bNm1jr7e2tQ4cO1enTpyftXb16db397W+v3t7euu666+rjH/94bd++vbq7u897761bt1ZnZ+fE1dPT0/yzAQAAgCtIU9E9MjJSbW1tk9ZaW1urqurMmTOT1n/+85/XbbfdVoODgzU6OlqPPvporV27tr73ve+d997r16+voaGhievw4cPNjAYAAABXnKaiu62trUZHRyetvf549uzZk9bvvffeuvXWW+uOO+6oWbNm1T333FNLly6t7du3n/feLS0t1dHRMekCAACAq1lT0d3X11cnT56s48ePT6wdOHCguru7q7Ozc9LeQ4cO1djY2KS16667rmbNmnUJ4wIAAMDVo6noXrhwYS1btqz6+/trZGSkXn311dqyZUutXbv2nL0f/vCH66GHHqrvfve7dfbs2dq9e3c9++yztWrVqikbHgAAAK5k5/9T4r/C7t276957760FCxbUNddcU3/yJ39SGzdurKqq9vb2euSRR+qTn/xkbdq0qa699tpasWJFnTp1qhYuXFh79+6t2267baqfAwAAAFyRZjQajcblHuJ8hoeHq7Ozs4aGhvx+NwAAANNiqlu0qbeXAwAAABdOdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAkKaj+8SJE7V8+fLq6uqquXPnVn9/f42Pj59373PPPVdLliyp9vb26unpqa1bt17ywAAAAHC1aDq6V61aVe3t7XX06NEaHBysffv21bZt287Z9+KLL9YHPvCBWrduXY2MjNRTTz1VDzzwQO3evXtKBgcAAIAr3YxGo9G40M0HDx6shQsX1pEjR2revHlVVfX444/X5z//+frxj388ae9f/MVf1KlTp2rnzp0Tay+99FJ1dHTUDTfccM69x8bGamxsbOLx8PBw9fT01NDQUHV0dDT9xAAAAKBZw8PD1dnZOWUt2tQr3fv37685c+ZMBHdVVW9vbx06dKhOnz49ae/g4GDdeOONtXr16po7d24tWrSoBgYGzhvcVVVbt26tzs7Oiaunp6f5ZwMAAABXkKaie2RkpNra2iattba2VlXVmTNnJq2fOnWqvvzlL9enPvWpOnbsWD3yyCP1uc997g3fXr5+/foaGhqauA4fPtzMaAAAAHDFaSq629raanR0dNLa649nz549ab2lpaU+8pGP1Ac/+MGaOXNmvfe97601a9bUrl27znvvlpaW6ujomHQBAADA1ayp6O7r66uTJ0/W8ePHJ9YOHDhQ3d3d1dnZOWlvb2/vpN/Rrqp67bXXqolfIQcAAICrWlPRvXDhwlq2bFn19/fXyMhIvfrqq7Vly5Zau3btOXs//elP1969e+vrX/96NRqNev7552vnzp21Zs2aKRseAAAArmRNf2TY7t27a3x8vBYsWFBLliypO++8szZu3FhVVe3t7RN/rfwP//AP64knnqgHH3ywOjs765577ql/+Id/qA9/+MNT+wwAAADgCtXUR4ZNp6n+M+0AAADw61zWjwwDAAAALpzoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAICQpqP7xIkTtXz58urq6qq5c+dWf39/jY+P/8qv+f73v1+tra01MDBwsXMCAADAVafp6F61alW1t7fX0aNHa3BwsPbt21fbtm17w/2jo6O1evXq+ulPf3pJgwIAAMDVpqnoPnjwYA0MDNT9999fra2tddNNN9XGjRvroYceesOvWbduXX30ox+95EEBAADgatNUdO/fv7/mzJlT8+bNm1jr7e2tQ4cO1enTp8/Z/7Wvfa0OHjxYmzZt+rX3Hhsbq+Hh4UkXAAAAXM2aiu6RkZFqa2ubtNba2lpVVWfOnJm0/uKLL9aGDRvqscceq2uvvfbX3nvr1q3V2dk5cfX09DQzGgAAAFxxmorutra2Gh0dnbT2+uPZs2dPrP3sZz+rVatW1Ze+9KWaP3/+Bd17/fr1NTQ0NHEdPny4mdEAAADgitNUdPf19dXJkyfr+PHjE2sHDhyo7u7u6uzsnFj7zne+Uy+99FKtXbu2urq6qqurq6qqPvShD9W6devOe++Wlpbq6OiYdAEAAMDVbEaj0Wg08wXvec97qru7ux599NH6v//7v/rjP/7j+tjHPlZ/+7d/+6u/0YwZ9eyzz9bv//7vX9D3GR4ers7OzhoaGhLgAAAATIupbtGmPzJs9+7dNT4+XgsWLKglS5bUnXfeWRs3bqyqqvb29tq5c+clDwUAAAC/CZp+pXu6eKUbAACA6XbZX+kGAAAALozoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhTUf3iRMnavny5dXV1VVz586t/v7+Gh8fP+/ehx9+uG655ZaaPXt23XzzzfWVr3zlkgcGAACAq0XT0b1q1apqb2+vo0eP1uDgYO3bt6+2bdt2zr69e/fW+vXra8eOHTU8PFw7duyoDRs21J49e6ZkcAAAALjSNRXdBw8erIGBgbr//vurtbW1brrpptq4cWM99NBD5+w9evRo/fVf/3X97u/+bs2YMaOWLl1af/AHf1DPP//8lA0PAAAAV7KZzWzev39/zZkzp+bNmzex1tvbW4cOHarTp09XV1fXxPq6desmfe2JEyfq+eefr3/8x388773HxsZqbGxs4vHw8HAzowEAAMAVp6lXukdGRqqtrW3SWmtra1VVnTlz5g2/7tixY3XXXXfV7bffXnffffd592zdurU6Ozsnrp6enmZGAwAAgCtOU9Hd1tZWo6Ojk9Zefzx79uzzfs23v/3tuuOOO+qWW26pJ554ombOPP+L6+vXr6+hoaGJ6/Dhw82MBgAAAFecpqK7r6+vTp48WcePH59YO3DgQHV3d1dnZ+c5+7/61a/W+973vurv76/HHnusWlpa3vDeLS0t1dHRMekCAACAq1lT0b1w4cJatmxZ9ff318jISL366qu1ZcuWWrt27Tl79+zZU3/+539e//qv/1qf/exnp2xgAAAAuFo0/ZFhu3fvrvHx8VqwYEEtWbKk7rzzztq4cWNVVbW3t9fOnTurqmrz5s01Pj5eK1asqPb29onr05/+9NQ+AwAAALhCzWg0Go3LPcT5DA8PV2dnZw0NDXmrOQAAANNiqlu06Ve6AQAAgAsjugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABAiOgGAACAENENAAAAIaIbAAAAQkQ3AAAAhIhuAAAACBHdAAAAECK6AQAAIER0AwAAQIjoBgAAgBDRDQAAACGiGwAAAEJENwAAAISIbgAAAAgR3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAgRHQDAABASNPRfeLEiVq+fHl1dXXV3Llzq7+/v8bHx8+79+mnn67FixdXW1tbLVq0qJ588slLHhgAAACuFk1H96pVq6q9vb2OHj1ag4ODtW/fvtq2bds5+15++eVasWJFbdmypYaGhmrz5s21cuXKOnLkyJQMDgAAAFe6GY1Go3Ghmw8ePFgLFy6sI0eO1Lx586qq6vHHH6/Pf/7z9eMf/3jS3i984Qs1ODhY//Ef/zGxdtddd9Xv/M7v1ObNm8+599jYWI2NjU08Hhoaqvnz59fhw4ero6Oj6ScGAAAAzRoeHq6enp46ffp0dXZ2XvL9Zjazef/+/TVnzpyJ4K6q6u3trUOHDtXp06erq6tr0t7FixdP+vre3t564YUXznvvrVu3njfGe3p6mhkRAAAALtnJkyenP7pHRkaqra1t0lpra2tVVZ05c2ZSdL/R3jNnzpz33uvXr6/PfOYzE49Pnz5dv/Vbv1WHDh2akicKV6LX/yuad3Twm85Z583AOefNwDnnzeD1d13PmTNnSu7XVHS3tbXV6OjopLXXH8+ePfuC9v7yvte1tLRUS0vLOeudnZ3+B81vvI6ODuecNwVnnTcD55w3A+ecN4NrrpmaD/tq6i59fX118uTJOn78+MTagQMHqru7+5xXo/v6+mr//v2T1g4cOFB9fX2XMC4AAABcPZqK7oULF9ayZcuqv7+/RkZG6tVXX60tW7bU2rVrz9m7Zs2aGhgYqF27dtX4+Hjt2rWrBgYGas2aNVM2PAAAAFzJmn69fPfu3TU+Pl4LFiyoJUuW1J133lkbN26sqqr29vbauXNnVVW9853vrL1799Z9991Xb3nLW+qLX/xi7dmzp26++eYL+j4tLS21adOm877lHH5TOOe8WTjrvBk457wZOOe8GUz1OW/qI8MAAACACzc1vxkOAAAAnEN0AwAAQIjoBgAAgBDRDQAAACGXNbpPnDhRy5cvr66urpo7d2719/fX+Pj4efc+/fTTtXjx4mpra6tFixbVk08+Oc3TwsVp5pw//PDDdcstt9Ts2bPr5ptvrq985SvTPC1cvGbO+uu+//3vV2traw0MDEzPkHCJmjnnzz33XC1ZsqTa29urp6entm7dOs3TwsVp5pw/+OCDtWDBguro6Kh3vetdtWfPnmmeFi7NT37yk3rHO97xK/8tcqktelmje9WqVdXe3l5Hjx6twcHB2rdvX23btu2cfS+//HKtWLGitmzZUkNDQ7V58+ZauXJlHTly5DJMDc250HO+d+/eWr9+fe3YsaOGh4drx44dtWHDBv/nxVXjQs/660ZHR2v16tX105/+dBqnhEtzoef8xRdfrA984AO1bt26GhkZqaeeeqoeeOCB2r1792WYGppzoef8m9/8Zt133331zDPP1PDwcG3atKlWrlxZP/rRj6Z/aLgI3/rWt2rp0qX1yiuvvOGeKWnRxmXy8ssvN6qqceTIkYm1b3zjG4358+efs3fDhg2NP/qjP5q0dueddzb+5m/+Jj4nXIpmzvk//dM/Nf7+7/9+0tpHP/rRxl/+5V/G54RL1cxZf92f/umfNjZu3Nioqsazzz47DVPCpWnmnN97772Nu+++e9LaD3/4w8b//u//xueES9HMOX/ggQcab3vb2xoHDhxonD17tvFv//Zvjeuvv75x+PDh6RwZLsr27dsb8+fPb3zjG9/4lf8WmYoWvWyvdO/fv7/mzJlT8+bNm1jr7e2tQ4cO1enTp8/Zu3jx4klrvb299cILL0zHqHDRmjnn69atq7/6q7+aeHzixIl6/vnn6/bbb5+uceGiNXPWq6q+9rWv1cGDB2vTpk3TOCVcmmbO+eDgYN144421evXqmjt3bi1atKgGBgbqhhtumOapoTnNnPPVq1fX29/+9urt7a3rrruuPv7xj9f27duru7t7mqeG5r3//e+vV155pVatWvUr901Fi1626B4ZGam2trZJa62trVVVdebMmQva+8v74ErTzDn/fx07dqzuuuuuuv322+vuu++OzghToZmz/uKLL9aGDRvqscceq2uvvXbaZoRL1cw5P3XqVH35y1+uT33qU3Xs2LF65JFH6nOf+5y3l3PFa+ac//znP6/bbrutBgcHa3R0tB599NFau3Ztfe9735u2eeFi3XDDDTVz5sxfu28qWvSyRXdbW1uNjo5OWnv98ezZsy9o7y/vgytNM+f8dd/+9rfrjjvuqFtuuaWeeOKJC/phAJfbhZ71n/3sZ7Vq1ar60pe+VPPnz5/WGeFSNfMzvaWlpT7ykY/UBz/4wZo5c2a9973vrTVr1tSuXbumbV64GM2c83vvvbduvfXWuuOOO2rWrFl1zz331NKlS2v79u3TNS7ETUWLXrbo7uvrq5MnT9bx48cn1g4cOFDd3d3V2dl5zt79+/dPWjtw4ED19fVNy6xwsZo551VVX/3qV+t973tf9ff312OPPVYtLS3TOS5ctAs969/5znfqpZdeqrVr11ZXV1d1dXVVVdWHPvShWrdu3XSPDU1p5md6b29vjY2NTVp77bXXqtFoTMuscLGaOeeHDh0655xfd911NWvWrGmZFabDlLToJfzu+SVbtmxZ4xOf+ERjeHi48T//8z+NW2+9tbFp06Zz9v3gBz9oXH/99Y3HH3+88Ytf/KLx+OOPN66//vrGD3/4w+kfGpp0oed89+7djVmzZjWeeeaZ6R8SpsCFnvVfVv6QGleRCz3n//mf/9mYOXNm45//+Z8bZ8+ebTz33HON9vb2xr//+79P/9DQpAs951/4whcac+fObfzXf/1X47XXXmv8y7/8S+P6669v/Pd///e0zwyX4lf9W2QqWvSyRvexY8caH/vYxxpvfetbG29729san/3sZxvj4+ONRqPRaGtra3z961+f2PvMM880fvu3f7vR3t7euPXWWxtPPfXU5RobmnKh53zx4sWNa665ptHW1jbp+rM/+7PLOT5csGZ+pv+/RDdXk2bO+dNPP91497vf3Zg9e3bjpptuajz88MOXa2xoyoWe81/84heNTZs2NW688cZGR0dH4/bbb/fiAVelX/63yFS36Iz//5sAAAAAU+yy/U43AAAA/KYT3QAAABAiugEAACBEdAMAAECI6AYAAIAQ0Q0AAAAhohsAAABCRDcAAACEiG4AAAAIEd0AAAAQIroBAAAg5P8DmDCt008DTfQAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1000x1000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Curious about the distribution of overlap sizes... \n",
    "\n",
    "\n",
    "def figure(ref_df:pd.DataFrame, path:str=None, min_overlap_fraction:float=0.8):\n",
    "    # ref_df = load_ref(genome_ids=genome_ids)\n",
    "    ref_df = ref_df[ref_df.spurious | ref_df.real].copy()\n",
    "\n",
    "    fig = plt.figure(figsize=(10, 10), layout='tight')\n",
    "    grid = GridSpec(nrows=1, ncols=1, figure=fig)\n",
    "    ax_a = fig.add_subplot(grid[0, 0])\n",
    "\n",
    "    ax_a_df = pd.DataFrame(index=ref_df.index)\n",
    "    ax_a_df['length'] = get_lengths(ref_df, top_hit=False, units='nt')\n",
    "    ax_a_df['overlap_length'] = ref_df['top_hit_overlap_length']\n",
    "    ax_a_df['overlap_fraction'] = ax_a_df.overlap_length / ax_a_df.length # This is relative to the query sequence. \n",
    "    ax_a_df['label'] = np.select([ref_df.real, ref_df.spurious], ['real', 'spurious'], default='none')\n",
    "    if min_overlap_fraction is not None:\n",
    "        ax_a_df = ax_a_df[ax_a_df.overlap_fraction > min_overlap_fraction].copy()\n",
    "\n",
    "    colors = ['darkseagreen', 'indianred']\n",
    "    # sns.kdeplot(data=ax_a_df, ax=ax_a, hue='label', hue_order=['real', 'spurious'], palette=colors, common_norm=False, x='overlap_fraction')\n",
    "    sns.ecdfplot(data=ax_a_df, ax=ax_a, hue='label', hue_order=['real', 'spurious'], palette=colors, x='overlap_fraction')\n",
    "\n",
    "    plt.show()\n",
    "    return ax_a_df\n",
    "\n",
    "\n",
    "ax_a_df = figure(ref_df, min_overlap_fraction=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tripy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
