{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.tools import MMseqs, NCBIDatasets, Prodigal\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from src.files import XMLFile, InterProScanFile, FASTAFile, GBFFFile\n",
    "from utils import * \n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split, GroupShuffleSplit\n",
    "from src.classifier import Classifier\n",
    "from src.sampler import Sampler\n",
    "from src.genome import ReferenceGenome\n",
    "from src.dataset import Dataset\n",
    "import re\n",
    "from src.build import * \n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phylum count: 56\n",
      "Class count: 123\n",
      "Order count: 299\n",
      "Genus count: 3983\n"
     ]
    }
   ],
   "source": [
    "ncbi_genome_metadata_df = load_ncbi_genome_metadata()\n",
    "\n",
    "for level in ['phylum', 'class', 'order', 'genus']:\n",
    "    print(f'{level.capitalize()} count:', ncbi_genome_metadata_df[f'{level}'].nunique())\n",
    "    # print(f'Species with no {level}:', (genome_metadata_df[f'{level}_taxid'] == 'none').sum())\n",
    "\n",
    "# fig, (ax_a, ax_b) = plt.subplots(ncols=2, figsize=(15, 5), width_ratios=[0.3, 0.7])\n",
    "\n",
    "# counts, bins, _ = ax_a.hist(ncbi_genome_metadata_df.gc_percent, bins=25, width=2, edgecolor='black', color='lightgray')\n",
    "# ax_a.set_ylabel('density')\n",
    "# ax_a.set_xlabel('GC percent')\n",
    "\n",
    "# level = 'phylum'\n",
    "# # x = np.arange(genome_metadata_df[f'{level}_taxid'].nunique())\n",
    "# y = ncbi_genome_metadata_df.groupby(level).gc_percent.mean() # .sort_values(ascending=True)\n",
    "# x = ncbi_genome_metadata_df.groupby(level).apply(len, include_groups=False) # .sort_values(ascending=True)\n",
    "# y_err = ncbi_genome_metadata_df.groupby(level).apply(lambda df : df.gc_percent.std() / np.sqrt(len(df)), include_groups=False) # .loc[y.index]\n",
    "\n",
    "# ax_b.errorbar(x, y, yerr=y_err, ls='', lw=0.7, capsize=2, color='black')\n",
    "# ax_b.scatter(x, y, color='black', s=5)\n",
    "# ax_b.set_xscale('log')\n",
    "# ax_b.set_xlabel(f'log({level} size)')\n",
    "# ax_b.set_ylabel('GC percent')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Low-GC content phylum: Campylobacterota\n",
      "Medium-GC content phylum: Planctomycetota\n",
      "High-GC content phylum: Deinococcota\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Start by selecting a few phyla in a reasonable size range with varying GC contents.\n",
    "def sample(genome_metadata_df:pd.DataFrame, min_phylum_size:int=90, max_phylum_size:int=200):\n",
    "\n",
    "    phylum_sizes = genome_metadata_df.groupby('phylum').apply(len, include_groups=False)\n",
    "    phylum_sizes = phylum_sizes[(phylum_sizes > min_phylum_size) & (phylum_sizes < max_phylum_size)]\n",
    "\n",
    "    genome_metadata_df = genome_metadata_df[genome_metadata_df.phylum.isin(phylum_sizes.index)]\n",
    "    \n",
    "    phylum_gc_percent = genome_metadata_df.groupby('phylum').gc_percent.mean() \n",
    "    phylum_gc_percent = phylum_gc_percent.sort_values(ascending=True)\n",
    "    \n",
    "    phyla = dict()\n",
    "    phyla['low_gc_content'] = phylum_gc_percent.index[0]\n",
    "    phyla['med_gc_content'] = phylum_gc_percent.index[len(phylum_gc_percent) // 2]\n",
    "    phyla['high_gc_content'] = phylum_gc_percent.index[-1]\n",
    "\n",
    "    print('Low-GC content phylum:', phyla['low_gc_content'])\n",
    "    print('Medium-GC content phylum:', phyla['med_gc_content'])\n",
    "    print('High-GC content phylum:', phyla['high_gc_content'])\n",
    "\n",
    "    genome_ids = {category:genome_metadata_df.index[genome_metadata_df.phylum == phylum] for category, phylum in phyla.items()}\n",
    "    return genome_ids\n",
    "\n",
    "genome_ids = sample(ncbi_genome_metadata_df)\n",
    "\n",
    "# ncbi = NCBIDatasets(genome_dir='../data/genomes', gbff_dir='../data/proteins/ncbi')\n",
    "# ncbi.run(genome_ids=genome_ids['low_gc_content'], metadata_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>product</th>\n",
       "      <th>note</th>\n",
       "      <th>pseudo</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>inference</th>\n",
       "      <th>experiment</th>\n",
       "      <th>ribosomal_slippage</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>continuous</th>\n",
       "      <th>seq</th>\n",
       "      <th>translation_table</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>evidence_category</th>\n",
       "      <th>evidence_details</th>\n",
       "      <th>evidence_source</th>\n",
       "      <th>used_pgap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>thr operon leader peptide</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b0001</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_414542.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>fused aspartate kinase/homoserine dehydrogenase 1</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b0002</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_414543.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b0003</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_414544.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b0004</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_414545.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>DUF2502 domain-containing protein YaaX</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b0005</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_414546.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4637</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>putative inner membrane protein CreD</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b4400</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MLKSPLFWKMTSLFGAVLLLLIPIMLIRQVIVERADYRSDVEDAIR...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_418817.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4638</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>DNA-binding transcriptional dual regulator ArcA</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b4401</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVEDELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_418818.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4639</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>uncharacterized protein YjjY</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b4402</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MTKVRNCVLDALSINVNNIISLVVGTFPQDPTVSKTAVILTILTAT</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_418819.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4640</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>protein YtjE</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b4824</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MSRFIFYKIILSKQKDM</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>YP_010283921.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4641</th>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>putative tRNA/rRNA methyltransferase YjtD</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>b4403</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRITIILVAPARAENIGAAARAMKTMGFSDLRIVDSQAHLEPATRW...</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>NP_418820.1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4642 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     feature  contig_id                                            product  \\\n",
       "0        CDS  NC_000913                          thr operon leader peptide   \n",
       "1        CDS  NC_000913  fused aspartate kinase/homoserine dehydrogenase 1   \n",
       "2        CDS  NC_000913                                  homoserine kinase   \n",
       "3        CDS  NC_000913                                 threonine synthase   \n",
       "4        CDS  NC_000913             DUF2502 domain-containing protein YaaX   \n",
       "...      ...        ...                                                ...   \n",
       "4637     CDS  NC_000913               putative inner membrane protein CreD   \n",
       "4638     CDS  NC_000913    DNA-binding transcriptional dual regulator ArcA   \n",
       "4639     CDS  NC_000913                       uncharacterized protein YjjY   \n",
       "4640     CDS  NC_000913                                       protein YtjE   \n",
       "4641     CDS  NC_000913          putative tRNA/rRNA methyltransferase YjtD   \n",
       "\n",
       "      note  pseudo locus_tag inference experiment  ribosomal_slippage  strand  \\\n",
       "0     none   False     b0001      none       none               False       1   \n",
       "1     none   False     b0002      none       none               False       1   \n",
       "2     none   False     b0003      none       none               False       1   \n",
       "3     none   False     b0004      none       none               False       1   \n",
       "4     none   False     b0005      none       none               False       1   \n",
       "...    ...     ...       ...       ...        ...                 ...     ...   \n",
       "4637  none   False     b4400      none       none               False       1   \n",
       "4638  none   False     b4401      none       none               False      -1   \n",
       "4639  none   False     b4402      none       none               False       1   \n",
       "4640  none   False     b4824      none       none               False       1   \n",
       "4641  none   False     b4403      none       none               False       1   \n",
       "\n",
       "      ...  continuous                                                seq  \\\n",
       "0     ...        True                              MKRISTTITTTITITTGNGAG   \n",
       "1     ...        True  MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2     ...        True  MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...   \n",
       "3     ...        True  MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...   \n",
       "4     ...        True  MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...   \n",
       "...   ...         ...                                                ...   \n",
       "4637  ...        True  MLKSPLFWKMTSLFGAVLLLLIPIMLIRQVIVERADYRSDVEDAIR...   \n",
       "4638  ...        True  MQTPHILIVEDELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "4639  ...        True     MTKVRNCVLDALSINVNNIISLVVGTFPQDPTVSKTAVILTILTAT   \n",
       "4640  ...        True                                  MSRFIFYKIILSKQKDM   \n",
       "4641  ...        True  MRITIILVAPARAENIGAAARAMKTMGFSDLRIVDSQAHLEPATRW...   \n",
       "\n",
       "     translation_table  codon_start      protein_id evidence_type  \\\n",
       "0                   11            1     NP_414542.1          none   \n",
       "1                   11            1     NP_414543.1          none   \n",
       "2                   11            1     NP_414544.1          none   \n",
       "3                   11            1     NP_414545.1          none   \n",
       "4                   11            1     NP_414546.1          none   \n",
       "...                ...          ...             ...           ...   \n",
       "4637                11            1     NP_418817.1          none   \n",
       "4638                11            1     NP_418818.1          none   \n",
       "4639                11            1     NP_418819.1          none   \n",
       "4640                11            1  YP_010283921.1          none   \n",
       "4641                11            1     NP_418820.1          none   \n",
       "\n",
       "     evidence_category evidence_details evidence_source used_pgap  \n",
       "0                 none             none            none     False  \n",
       "1                 none             none            none     False  \n",
       "2                 none             none            none     False  \n",
       "3                 none             none            none     False  \n",
       "4                 none             none            none     False  \n",
       "...                ...              ...             ...       ...  \n",
       "4637              none             none            none     False  \n",
       "4638              none             none            none     False  \n",
       "4639              none             none            none     False  \n",
       "4640              none             none            none     False  \n",
       "4641              none             none            none     False  \n",
       "\n",
       "[4642 rows x 23 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = GBFFFile('../data/ncbi/gbffs/GCF_000005845.2_genomic.gbff')\n",
    "df = file.to_df()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_metadata_df = list()\n",
    "for category, genome_ids_ in genome_ids.items():\n",
    "    df = ncbi_genome_metadata_df.loc[genome_ids_, :].copy()\n",
    "    df['gc_percent_category'] = category\n",
    "    # df = df[genome_metadata_df.annotation_pipeline.str.contains('PGAP')] # Only include genomes annotated with PGAP, because the other files are not consistent. \n",
    "    genome_metadata_df.append(df)\n",
    "genome_metadata_df = pd.concat(genome_metadata_df)\n",
    "genome_metadata_df.to_csv('../data/genome_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi = NCBIDatasets()\n",
    "ncbi.run(genome_ids=model_organism_genome_ids, metadata=False)\n",
    "ncbi.cleanup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/prichter/Documents/tripy/src/__init__.py:41: UserWarning: get_dtypes: Column \"source\" only contains NaNs. Inferring datatype as strings.\n",
      "  warnings.warn(f'get_dtypes: Column \"{col}\" only contains NaNs. Inferring datatype as strings.')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<src.files.gbff.GBFFFile at 0x72e545da0e90>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GBFFFile('../data/ncbi/gbffs/GCF_000006765.1_genomic.gbff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "prodigal = Prodigal()\n",
    "for genome_id in model_organism_genome_ids:\n",
    "    input_path = f'../data/ncbi/genomes/{genome_id}_genomic.fna'\n",
    "    output_path = f'../data/proteins/{genome_id}_protein.faa'\n",
    "    if not os.path.exists(output_path):\n",
    "        prodigal.run(input_path, output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome_id in model_organism_genome_ids:\n",
    "    ! ref --input-path ../data/proteins/{genome_id}_protein.faa --output-dir ../data/ref --gbffs-dir ../data/ncbi/gbffs/ --prodigal-output --summarize\n",
    "# # ! ref --input-path ../data/proteins/GCF_003711085.1_protein.faa --overwrite --output-dir ../data/ref --gbffs-dir ../data/ncbi/gbffs/ --prodigal-output --summarize "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = load_ref(genome_ids=[genome_id for genome_id in genome_ids['low_gc_content'] if (genome_id != 'GCF_000009085.1')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "top_hit_genome_id GCF_000259275.1 391\n",
      "query_strand -1 top_hit_strand -1\n",
      "query_seq (402789, 403010) MDFGQKWLGVSVSQNYKNAIKPLFETLEKAKKEGMLWRNFSNKEQEIYAPLLQAFKDEVLRIDENKKNKVPQK\n",
      "top_hit_seq (402451, 403402, codon_start=1) none\n",
      "overlap_length (402789, 403010) 222\n",
      "query_overlap_fraction 1.0 top_hit_overlap_fraction 0.2331932773109243\n",
      "top_hit_product HaeIII family restriction endonuclease\n",
      "n_hits 1\n",
      "top_hit_note frameshifted; Derived by automated computational analysis using gene prediction method: Protein Homology.\n"
     ]
    }
   ],
   "source": [
    "def info(ref_df:pd.DataFrame, id_:str):\n",
    "    row = ref_df.loc[id_]\n",
    "    print('top_hit_genome_id', row.top_hit_genome_id, f'{row.top_hit_id}')\n",
    "    print('query_strand', row.query_strand, 'top_hit_strand', row.top_hit_strand)\n",
    "    print('query_seq', f'({row.query_start}, {row.query_stop})', row.query_seq)\n",
    "    print('top_hit_seq', f'({row.top_hit_start}, {row.top_hit_stop}, codon_start={row.top_hit_codon_start})', row.top_hit_seq)\n",
    "    print('overlap_length', f'({row.overlap_start}, {row.overlap_stop})', row.overlap_length)\n",
    "    print('query_overlap_fraction', row.query_overlap_fraction, 'top_hit_overlap_fraction', row.top_hit_overlap_fraction)\n",
    "    print('top_hit_product', row.top_hit_product)\n",
    "    print('n_hits', row.n_hits)\n",
    "    print('top_hit_note', row.top_hit_note)\n",
    "\n",
    "\n",
    "info(ref_df, 'NC_017735.1_382')\n",
    "# 443855"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "build_datasets: Loading data from 196 genomes belonging to the phylum Campylobacterota.\n",
      "Removing 383 sequences exceeding the maximum length of 2000\n",
      "build_datasets: Loaded 357263 sequences, 297 labeled spurious and 356966 labeled real.\n",
      "MMseqs.load: Removing 300625 non-cluster representatives.\n",
      "build_datasets: Clustering at 50 percent similarity removed 300625 sequences.\n",
      "build_datasets: 240 negative instances and 43785 positive instances in the training dataset.\n",
      "build_datasets: 57 negative instances and 12556 positive instances in the testing dataset.\n"
     ]
    }
   ],
   "source": [
    "def build_datasets(genome_metadata_df:pd.DataFrame, phylum:str='Campylobacterota', max_length:int=2000):\n",
    "    # First remove things which do not have definitive labels. \n",
    "    genome_ids = genome_metadata_df[genome_metadata_df.phylum == phylum].index\n",
    "    genome_ids = [genome_id for genome_id in genome_ids if (genome_id != 'GCF_000009085.1')] # Remove the non-PGAP genome. \n",
    "\n",
    "    print(f'build_datasets: Loading data from {len(genome_ids)} genomes belonging to the phylum {phylum}.')\n",
    "    df = load_ref(genome_ids=genome_ids, add_labels=True)\n",
    "    df = df.rename(columns={'query_seq':'seq'}) # Need to do this for file writing, etc. to work correctly, \n",
    "    df = df.drop(columns=['top_hit_homolog_id', 'top_hit_homolog_seq', 'pseudo'])\n",
    "\n",
    "    lengths = df.seq.apply(len)\n",
    "    print(f'Removing {(lengths >= max_length).sum()} sequences exceeding the maximum length of {max_length}')\n",
    "    df = df[lengths < max_length]\n",
    "    all_df = df.copy()\n",
    "\n",
    "    df = df[df.label != 'none'].copy()\n",
    "    df.label = [0 if (label == 'spurious') else 1 for label in df.label]\n",
    "\n",
    "    print(f'build_datasets: Loaded {len(df)} sequences, {(df.label == 0).sum()} labeled spurious and {(df.label == 1).sum()} labeled real.')\n",
    "\n",
    "    real_df, spurious_df = df[df.label == 1].copy(), df[df.label == 0].copy()\n",
    "    n_real = len(real_df)\n",
    "\n",
    "    mmseqs = MMseqs()\n",
    "    real_df = mmseqs.cluster(real_df, job_name=f'{phylum.lower()}_real_50', sequence_identity=0.50, reps_only=True, overwrite=False)\n",
    "    print(f'build_datasets: Clustering at 50 percent similarity removed {n_real - len(real_df)} sequences.')\n",
    "    mmseqs.cleanup()\n",
    "\n",
    "    df = pd.concat([spurious_df, real_df], ignore_index=False)\n",
    "\n",
    "    gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "    train_idxs, test_idxs = list(gss.split(df, groups=df.genome_id))[0]\n",
    "    train_df, test_df = df.iloc[train_idxs], df.iloc[test_idxs]\n",
    "    print(f'build_datasets: {(train_df.label == 0).sum()} negative instances and {(train_df.label == 1).sum()} positive instances in the training dataset.')\n",
    "    print(f'build_datasets: {(test_df.label == 0).sum()} negative instances and {(test_df.label == 1).sum()} positive instances in the testing dataset.')\n",
    "\n",
    "    all_df['in_test_dataset'] = all_df.index.isin(test_df.index)\n",
    "    all_df['in_train_dataset'] = all_df.index.isin(train_df.index)\n",
    "    \n",
    "    train_df.to_csv(f'../data/{phylum.lower()}_dataset_train.csv')\n",
    "    test_df.to_csv(f'../data/{phylum.lower()}_dataset_test.csv')\n",
    "    all_df.to_csv(f'../data/{phylum.lower()}_dataset.csv')\n",
    "\n",
    "    return train_df, test_df, all_df \n",
    "    \n",
    "train_df, test_df, all_df = build_datasets(genome_metadata_df)\n",
    "# phylum ='Campylobacterota'\n",
    "# dtypes = {'top_hit_partial':str, 'query_partial':str, 'top_hit_translation_table':str, 'top_hit_codon_start':str}\n",
    "# train_df = pd.read_csv(f'../data/{phylum.lower()}_dataset_train.csv', index_col=0, dtype=dtypes)\n",
    "# test_df = pd.read_csv(f'../data/{phylum.lower()}_dataset_test.csv', index_col=0, dtype=dtypes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000005845.2 101\n",
      "GCF_000006765.1 74\n",
      "GCF_000009045.1 28\n",
      "GCF_000195955.2 108\n"
     ]
    }
   ],
   "source": [
    "# I want to get a rough estimate of how many spurious sequences we can expect in a genome, which I think I will need to use\n",
    "# model organisms for. I need genomes which do not have suspect genes in the reference, i.e. those which are not based solely on ab initio prediction\n",
    "# and have no annotation. \n",
    "ref_df = load_ref(genome_ids=model_organism_genome_ids, add_labels=False)\n",
    "\n",
    "for genome_id, df in ref_df.groupby('genome_id'):\n",
    "    print(genome_id, (df.n_hits == 0).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = load_ref(genome_ids=model_organism_genome_ids, add_labels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Balanced accuracy: 0.8372044119537454\n",
      "Precision (class 0): 0.0004276702662247407\n",
      "Recall (class 0): 0.07017543859649122\n",
      "Number of false negatives: 3207\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'confusion_matrix'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_2132789/38299336.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0mconfusion_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwhere\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodel_label\u001b[0m  \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlabel\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'true negative'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconfusion_matrix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'confusion_matrix'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconfusion_matrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0mbenchmark_fn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconfusion_matrix\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'false negative'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m \u001b[0mfn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpred_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpred_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconfusion_matrix\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'false negative'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintersect1d\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbenchmark_fn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/tripy/lib/python3.13/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6295\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6296\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6297\u001b[0m         \u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6298\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6299\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'confusion_matrix'"
     ]
    }
   ],
   "source": [
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import balanced_accuracy_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "class BenchmarkClassifier():\n",
    "\n",
    "    def __init__(self, feature_col:str='query_length'):\n",
    "\n",
    "        self.model = LogisticRegression(class_weight='balanced', C=1) # , multi_class='multinomial') \n",
    "        self.scaler = StandardScaler()\n",
    "        self.feature_col = feature_col\n",
    "\n",
    "    def predict(self, df:pd.DataFrame, include_outputs:bool=True):\n",
    "        X = df[[self.feature_col]].values\n",
    "        X = self.scaler.transform(X)\n",
    "        model_labels = self.model.predict(X)\n",
    "        model_outputs = self.model.predict_proba(X)\n",
    "\n",
    "        return model_labels if (not include_outputs) else (model_labels, model_outputs)\n",
    "\n",
    "    def fit(self, df:pd.DataFrame):\n",
    "        X, labels = df[[self.feature_col]].values, df['label'].values\n",
    "        X = self.scaler.fit_transform(X)\n",
    "        self.model.fit(X, labels)\n",
    "\n",
    "model = BenchmarkClassifier()\n",
    "model.fit(train_df)\n",
    "\n",
    "df = test_df[['label', 'query_length']].copy()\n",
    "df['model_label'], outputs = model.predict(test_df, include_outputs=True)\n",
    "df['model_output_0'], df['model_output_1'] = outputs[:, 0].ravel(), outputs[:, 1].ravel()\n",
    "\n",
    "print('Balanced accuracy:', balanced_accuracy_score(df.label.values, df.model_label.values))\n",
    "print('Precision (class 0):', precision(df, class_=0))\n",
    "print('Recall (class 0):', recall(df, class_=0))\n",
    "print('Number of false negatives:', ((df.label == 1) & (df.model_label == 0)).sum())\n",
    "\n",
    "\n",
    "confusion_matrix = np.where((df.model_label == 1) & (df.label == 0), 'false positive', '')\n",
    "confusion_matrix = np.where((df.model_label  == 1) & (df.label == 1), 'true positive', confusion_matrix)\n",
    "confusion_matrix = np.where((df.model_label == 0) & (df.label == 1), 'false negative', confusion_matrix)\n",
    "confusion_matrix = np.where((df.model_label  == 0) & (df.label == 0), 'true negative', confusion_matrix)\n",
    "df['confusion_matrix'] = confusion_matrix\n",
    "\n",
    "benchmark_fn = df[df.confusion_matrix == 'false negative'].index\n",
    "fn = pred_df[pred_df.confusion_matrix == 'false negative'].index\n",
    "np.intersect1d(fn, benchmark_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There still seems to be something wrong with the training data, perhaps I am still misclassifying the spurious sequences?\n",
    "# It is possible my \"in-frame\" check is wrong, it's possible I should only see if either edge is a multiple of 3 base pairs away, \n",
    "# as opposed to both.\n",
    "\n",
    "# It seems as though 684 of the sequences flagged as spurious have an overlap length of 0, so maybe I should be counting those as intergenic?\n",
    "# Ok, so I think because of how frequent gene overlap is, I should be allowing some (considering genes with less than a certain amount of overlap to be intergenic)\n",
    "# I should also be very conservative, i.e. allow a comparatively large amount of overlap. \n",
    "# https://pmc.ncbi.nlm.nih.gov/articles/PMC525685/ \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Curious about the distribution of overlap sizes... \n",
    "\n",
    "\n",
    "# def figure(ref_df:pd.DataFrame, path:str=None, min_overlap_fraction:float=0.8):\n",
    "#     # ref_df = load_ref(genome_ids=genome_ids)\n",
    "#     ref_df = ref_df[ref_df.spurious | ref_df.real].copy()\n",
    "\n",
    "#     fig = plt.figure(figsize=(10, 10), layout='tight')\n",
    "#     grid = GridSpec(nrows=1, ncols=1, figure=fig)\n",
    "#     ax_a = fig.add_subplot(grid[0, 0])\n",
    "\n",
    "#     ax_a_df = pd.DataFrame(index=ref_df.index)\n",
    "#     ax_a_df['length'] = get_lengths(ref_df, top_hit=False, units='nt')\n",
    "#     ax_a_df['overlap_length'] = ref_df['top_hit_overlap_length']\n",
    "#     ax_a_df['overlap_fraction'] = ax_a_df.overlap_length / ax_a_df.length # This is relative to the query sequence. \n",
    "#     ax_a_df['label'] = np.select([ref_df.real, ref_df.spurious], ['real', 'spurious'], default='none')\n",
    "#     if min_overlap_fraction is not None:\n",
    "#         ax_a_df = ax_a_df[ax_a_df.overlap_fraction > min_overlap_fraction].copy()\n",
    "\n",
    "#     colors = ['darkseagreen', 'indianred']\n",
    "#     # sns.kdeplot(data=ax_a_df, ax=ax_a, hue='label', hue_order=['real', 'spurious'], palette=colors, common_norm=False, x='overlap_fraction')\n",
    "#     sns.ecdfplot(data=ax_a_df, ax=ax_a, hue='label', hue_order=['real', 'spurious'], palette=colors, x='overlap_fraction')\n",
    "\n",
    "#     plt.show()\n",
    "#     return ax_a_df\n",
    "\n",
    "\n",
    "# ax_a_df = figure(ref_df, min_overlap_fraction=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tripy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
