{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.tools import MMseqs, NCBIDatasets, UniRef\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from src.files import XMLFile, InterProScanFile, FASTAFile, GBFFFile\n",
    "from utils import * \n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split, GroupShuffleSplit\n",
    "from src.classifier import Classifier\n",
    "from src.sampler import Sampler\n",
    "from src.genome import ReferenceGenome\n",
    "from src.dataset import Dataset\n",
    "import re\n",
    "from src.build import * \n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phylum count: 56\n",
      "Class count: 123\n",
      "Order count: 299\n",
      "Genus count: 3983\n"
     ]
    }
   ],
   "source": [
    "ncbi_genome_metadata_df = load_ncbi_genome_metadata()\n",
    "\n",
    "for level in ['phylum', 'class', 'order', 'genus']:\n",
    "    print(f'{level.capitalize()} count:', ncbi_genome_metadata_df[f'{level}'].nunique())\n",
    "    # print(f'Species with no {level}:', (genome_metadata_df[f'{level}_taxid'] == 'none').sum())\n",
    "\n",
    "# fig, (ax_a, ax_b) = plt.subplots(ncols=2, figsize=(15, 5), width_ratios=[0.3, 0.7])\n",
    "\n",
    "# counts, bins, _ = ax_a.hist(ncbi_genome_metadata_df.gc_percent, bins=25, width=2, edgecolor='black', color='lightgray')\n",
    "# ax_a.set_ylabel('density')\n",
    "# ax_a.set_xlabel('GC percent')\n",
    "\n",
    "# level = 'phylum'\n",
    "# # x = np.arange(genome_metadata_df[f'{level}_taxid'].nunique())\n",
    "# y = ncbi_genome_metadata_df.groupby(level).gc_percent.mean() # .sort_values(ascending=True)\n",
    "# x = ncbi_genome_metadata_df.groupby(level).apply(len, include_groups=False) # .sort_values(ascending=True)\n",
    "# y_err = ncbi_genome_metadata_df.groupby(level).apply(lambda df : df.gc_percent.std() / np.sqrt(len(df)), include_groups=False) # .loc[y.index]\n",
    "\n",
    "# ax_b.errorbar(x, y, yerr=y_err, ls='', lw=0.7, capsize=2, color='black')\n",
    "# ax_b.scatter(x, y, color='black', s=5)\n",
    "# ax_b.set_xscale('log')\n",
    "# ax_b.set_xlabel(f'log({level} size)')\n",
    "# ax_b.set_ylabel('GC percent')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Low-GC content phylum: Campylobacterota\n",
      "Medium-GC content phylum: Planctomycetota\n",
      "High-GC content phylum: Deinococcota\n"
     ]
    }
   ],
   "source": [
    "# How should I select the genomes for model training?\n",
    "\n",
    "# Using the same genomes for validation as in the Prodigal paper\n",
    "# Halobacterium salinarum\n",
    "# Natronomonas pharaonis\n",
    "# Aeropyrum pernix\n",
    "model_organism_genome_ids = ['GCF_000005845.2'] # E. coli K-12\n",
    "model_organism_genome_ids += ['GCF_000009045.1'] # B. subtilis\n",
    "model_organism_genome_ids += ['GCF_000006765.1'] # P. aeruginosa\n",
    "\n",
    "# Start by selecting a few phyla in a reasonable size range with varying GC contents.\n",
    "def sample(genome_metadata_df:pd.DataFrame, min_phylum_size:int=90, max_phylum_size:int=200):\n",
    "\n",
    "    phylum_sizes = genome_metadata_df.groupby('phylum').apply(len, include_groups=False)\n",
    "    phylum_sizes = phylum_sizes[(phylum_sizes > min_phylum_size) & (phylum_sizes < max_phylum_size)]\n",
    "\n",
    "    genome_metadata_df = genome_metadata_df[genome_metadata_df.phylum.isin(phylum_sizes.index)]\n",
    "    \n",
    "    phylum_gc_percent = genome_metadata_df.groupby('phylum').gc_percent.mean() \n",
    "    phylum_gc_percent = phylum_gc_percent.sort_values(ascending=True)\n",
    "    \n",
    "    phyla = dict()\n",
    "    phyla['low_gc_content'] = phylum_gc_percent.index[0]\n",
    "    phyla['med_gc_content'] = phylum_gc_percent.index[len(phylum_gc_percent) // 2]\n",
    "    phyla['high_gc_content'] = phylum_gc_percent.index[-1]\n",
    "\n",
    "    print('Low-GC content phylum:', phyla['low_gc_content'])\n",
    "    print('Medium-GC content phylum:', phyla['med_gc_content'])\n",
    "    print('High-GC content phylum:', phyla['high_gc_content'])\n",
    "\n",
    "    genome_ids = {category:genome_metadata_df.index[genome_metadata_df.phylum == phylum] for category, phylum in phyla.items()}\n",
    "    return genome_ids\n",
    "\n",
    "genome_ids = sample(ncbi_genome_metadata_df)\n",
    "\n",
    "# ncbi = NCBIDatasets(genome_dir='../data/genomes', gbff_dir='../data/proteins/ncbi')\n",
    "# ncbi.run(genome_ids=genome_ids['low_gc_content'], metadata_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_metadata_df = list()\n",
    "for category, genome_ids_ in genome_ids.items():\n",
    "    df = ncbi_genome_metadata_df.loc[genome_ids_, :].copy()\n",
    "    df['gc_percent_category'] = category\n",
    "    # df = df[genome_metadata_df.annotation_pipeline.str.contains('PGAP')] # Only include genomes annotated with PGAP, because the other files are not consistent. \n",
    "    genome_metadata_df.append(df)\n",
    "genome_metadata_df = pd.concat(genome_metadata_df)\n",
    "# genome_metadata_df.to_csv('../data/genome_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from src.tools import download_homologs\n",
    "# for path in glob.glob('../data/proteins/ncbi/*'):\n",
    "#     download_homologs(path, pseudo_only=True, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for genome_id in genome_ids['low_gc_content']:\n",
    "#     ! ref --input-path ../data/proteins/prodigal/{genome_id}_protein.faa --output-dir ../data/ref --reference-dir ../data/proteins/ncbi/ --prodigal-output --summarize --load-homologs --homologs-dir ../data/proteins/homologs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>md5</th>\n",
       "      <th>length</th>\n",
       "      <th>analysis</th>\n",
       "      <th>signature</th>\n",
       "      <th>signature_description</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>e_value</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>annotation</th>\n",
       "      <th>description</th>\n",
       "      <th>go_annotation</th>\n",
       "      <th>pathways_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, md5, length, analysis, signature, signature_description, start, stop, e_value, status, date, annotation, description, go_annotation, pathways_annotation]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = InterProScanFile('../data/interpro/GCF_013372265.1_interpro.tsv')\n",
    "file.df[(file.df.analysis == 'TIGERFAM') & (file.df.e_value == '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = load_ref(genome_ids=[genome_id for genome_id in genome_ids['low_gc_content'] if (genome_id != 'GCF_000009085.1')])\n",
    "\n",
    "# My approach to spurious sequences still appears to not be catching everything. \n",
    "# Is this possibly related to the fact that the translate function isn't working properly. \n",
    "\n",
    "# Translate function is OK, but frameshifts are causing problems. Example case is the sequence NC_008229.1_452, which is one of several\n",
    "# sequences in the bounds of a pseudogene with lots of internal stops. Basically, Prodigal has picked up on like 5 proteins, all of which are\n",
    "# sections of the frameshifted pseudogene. However, trying to translate the pseudogene doesn't result in anything that really makes sense, (it just \n",
    "# gets a few amino acids before the internal stop, and then there are a bunch more internal stops). Also, none of the sub-sequences are in-frame\n",
    "# with the labeled pseudogene (this gene doesn't even have a lenght divisible by three). So, all of the Prodigal proteins were registered as \n",
    "# spurious, but all bear homology to the pseudogene. I think I will need to handle validating pseudogenes separately. \n",
    "\n",
    "# I should also note that pseudogenes that are partial in the middle are not necessarily marked partial by their coordinates. Additionally, some\n",
    "# genes marked partial by their coordinates are not pseudo. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_spurious_df = ref_df[ref_df.spurious]\n",
    "ref_intergenic_df = ref_df[ref_df.spurious]\n",
    "FASTAFile(df=ref_intergenic_df.rename(columns={'query_seq':'seq'})).write('../data/campylobacterota_intergenic.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGAACCACAACAAACACACCATAAACACCGCAAAATAAAGCGCCCTATGGTTCCTCTCGCTTGAGCGGGAGCATTAATTAGCACTAAATACAACCCGCTCCCATCCATGTTGAAAATCGCGCCCTTTTATCAGGCGTGATTATTCTAGCCATCCTTAGTGGTATGATCGCAGGTGTCGCTTCAAGAACAGCCGGAGGGATTCTTGGTGGGGCGCTCAAACAAACCAATAAAATTTGGCGCATTCAAGCAGGGAGAGGCTTTGATTCTTCCCCTAACAAGCAATACGGCTTATACAAATCCCTATTGTCTAGCAACATTCAAAGAGGTTAGGATTGGGGGAATGCTGCTAGGCATGATTGGGTCAAAGGCTAGCAGCGCAATCAGTTTGAAGCGGGTATGAAAAACGCCGTAGGGGCTTATAACCTTTCAGGGCTCAGAAACTATACCGGTGGGGATTTAGATGCCAATATGCAAAAAGCCGCTTGGCGCTTGGGCCAATCCAATGGCAGCTCTTCCACAAACTTTAAAGAGGCTGCCAATCGCGCCACAAGGGTGAATTCCAACGCTAAAAATGTCATAATCGGTAATTTTTTAGAGATCCATAACCGCATGGGTTCTGGGGCTAGGAGAAAAGCCAGCTCTGCGATTTTAACTTTACAAAGTGCAAGAAGGGATTGCAAGCGCTGAAAACGCTGAAATTTCTCTTTATGATGGCACTACCCTTAATTTGCAAGGAGCGCGCGCTGTGAATTTATTGGGTAAGGTGTGGATGGGCCGTTTGCAATATGCGGGAGCGTATTTAGCCCCTTCATGGAGCATGGTGAAGCATCTTGGCATTTTAAATACCCACTACAATATTTCGCATTTGTTCTAGCGCCCGATAAACACTTCAAAGGTGCAAGGGGATGTGAATTGGTGCCATCTCGTTGTGGGTGATCACAACGCCGCTCAAGCAGGCATTGCTGCTAATGAAAAGACTGACATTGCGCGCTGGATTTGTGACAAAGCATAGGATTAAACCTTATTGCTTCTCCAGAAGGTGGCTATAAAAAGACCACTTCTCAAAATAACGCTGACAATACCCAACAAAATAGCAGCCAAACGAATAGCCAAAACCATGAATGGCGCTCAAGCCATTAACCCGCCCAATAGCGTGCAAAAAGTAGAAGTTGAACCCATGAAAGTCATTGACGGGACTTTTGCTGGCGCTAGGGGCACGGTCGTGAATATTTTCCACTGAAACACTGACGCTGATGGGACGATTAAAGCGGGAGGGCATAAAGCTTTTGTTAGCGCCAGTGCGGCTCATTTGAATGTCGGCGAAGGCGCTATCCATCTATCCAATTCAACGAGTGGGCGCGCTCTTTTAGCGGGAAATCTAACCGGGAATACCACCGTTAATGGGGCATTAAGGGTGAATAAAGAAGTGGGTGGTGCTTTGCCAGGTTCAAGCGCGAATTTTGAATTTAAAGTTGGTGTGGATACCCAAAAGGGTGCAGCCGCTTTTAATAACGATACCCGTTTGGGAAAAGCGGTGGATTTAAAAGCGGGTGTGCATGCAGTTAATTTTAATGGTGATGCGTATTGAGGAAAATTTACGCATTTTAAAGGGGGTGCCAATACCGCTAGTTTTAAAGATATTGATGCCAATAAGGGTGGTAATGGCATCAACGCTACCACTTTGGATTTTAGGAACGTTATTAGGCATGGCCAATACCAACAAGCACAGCTTCCACTCATGCTGCCATTGAAAACTTCAGTATTAAGGAATTGGTTGTTACAACCCATGGTATGAGTGCGGGAAAATGCTCTCCATTTAGCGGGGATATAGGCGATAAATCGCGCATTACTACCGTGCGTTTAGAAACGGGATATAGCCCGCTCTATGCTAGAGGGGTTAAATTTAAAAGCGGTAAAAAACCCGTTATAGATGGGATGTGCCATGCTCCTTGGAATTATTTTGACGCCAGAAACACTACCGGTGTTGAAACCACCAAGAAAATTCTTTTTGGAGCCCTAGGAAACCTTGCAGGCAAGATAGGGCTTAATAACCTAACCTTGGGTCATCATGCAAGCATGGATTATGGTAAAGATTTGGATTTAACCCTTCAGGGGCATTTTATTAACAATCAAGGCGTGATGGATCTTTTTGTTCAAAACTGGCGTGCAGCAACCTTGAATGCAAGCCATCAAGCAAGCATGCTGTTTAATAACATGACAGACAACACAACCGGATTTTACAAACCGCTCATTAAGATCAATAACGCTCAAAACCTCACTAAAAGCACAGAGCATGTTTTTGGTGGGAGCGCGAAACATTGATGATCATTGAGCGGGAGTGCAAGGCGTTAGTTATGATGCTGTTTTTGCAAGCAATGCTAATCTGCAAGAGCAATTCAAAGAGTGCCTAGCCCTTTATAACAACAATACCCGTATGGGTATTTGTGTGGCGCAAAAAGGCAACACCAACGGTATTAAAGCATGCGGAATGGCTATCGGTAATCATGACATGAACAGTAACCCTAACCATTACAAGTATCTTGAAGGTAAGGCATGGAAAAATACAGGCATTAGTCAAACAGCCAACAACGCTAAAACCTCTGTTCGTTTGGGTAACAATTCTATGCCTAGTGGGCATGACAACAATGCTGCAAATTGACCCACAAACAACAGACAGGCGCGTTTCCTACGCTCTCATAAAAAACGCTCCTTTGGTCAAACGCAAGCGCCTCTTCTAAAGCGCCACTCCTAATTTAGTCCACATCCATGAGCCTCATTTTGGCACCCTTGAGAGCGCGTTTGAATTGGCTAACCGCTCCAAAGATATTGACGCGCTTTATGCCCATTCAGGCACGCAAGGCAGGAATCTCTTGCAAACCTTATTGGTTGATCATCATGATATTAGGCTATGCCAGAGAAATGATTGATAACACAAGTGCCAGCGAAACCACCAAGCAATTGGATGCGGCAACTGACACTTGAAACAGCATAGCCCGTTTGGATCGCAAAACAAGTGGTTTGCAAACTTTGAGTTTGAGCAATGCGATGGTTTTGAATTCTCGTTTAGCCAATCTCTCCAGGAAGCGCACCAACAATATTGGCTCATTCGCTAAACGCTTGCAAGCTTTAAAAGACCAAAAATTTAGCCCCACAGAAAGCGCGGCAGAAGTGTTGTATAAGTTCGCTCCTAACCATGAAAAACTTAAAAATCTTTGGGCTAATGCTATCAAGGGAGCGAGTTTGAATAGTGGTGGCAACACTTCATTATAGGGCACAAGCGCAGGGGTGGATACTTGCATTAACGGGGAAGTGGAAGCTCTTGTTGACGGGTTTGGAAGCGATGGTTATAGCGCCTTCAGTAACCGCGCTAGCCCTCTTAACTCTGATGGCAATAACGCTCATTTTGGCGTGTATAACCGTATCTTGGCTGACCAACAAAAATTTGACTTTCAAGCTGGGCTGTGGAAAGCGATCAATCAAGCTTGAATTGCAAAGGCGCTTTGTGGCGCTATGGCTAACGCCAGCTATGGTTATGGCTTTGGGTTTTTCAAAAACGCCTTAGCGCTAAAACCAAGCGTGGGAGTGATTCATAATCATTTAGGTTCAACCAACTTTGAAAGCAATAGCGCCCAAAAGCGGCTTTAAGCGATGGTGCAAGCGGTCAGCATTTATCCAACGCTGACGGCAACCTACAAGCCCGCTATTATTGCCGAGACACTTCATGCTTCTGTTTGCATGCTGGAGTTTTACAAGAGTTTGCCAACTTTGAATCCAACCACGCTGTGTCTTTGAACGCCTTCAAAGCAAATGCCGCTCGCAACCCTTTAAACACGCATGCCAGAGTGATGATGGGTGGGGAATTGGATTTGGCTAAAGAAGTGTTTTTGAATTGGGGCTTCATTGATTTGGACAACTTGATTTCCAATGCGAGCCGTTTCGCTTCCAGTTGA\n",
      "ATGGAACCACAACAAACACACCATAAACACCGCAAAATAAAGCGCCCTATGGTTCCTCTCGCTTGAGCGGGAGCATTAATTAGCACTAAATACAACCCGCTCCCATCCATGTTGAAAATCGCGCCCTTTTATCAGGCGTGATTATTCTAGCCATCCTTAGTGGTATGATCGCAGGTGTCGCTTCAAGAACAGCCGGAGGGATTCTTGGTGGGGCGCTCAAACAAACCAATAAAATTTGGCGCATTCAAGCAGGGAGAGGCTTTGATTCTTCCCCTAACAAGCAATACGGCTTATACAAATCCCTATTGTCTAGCAACATTCAAAGAGGTTAGGATTGGGGGAATGCTGCTAGGCATGATTGGGTCAAAGGCTAGCAGCGCAATCAGTTTGAAGCGGGTATGAAAAACGCCGTAGGGGCTTATAACCTTTCAGGGCTCAGAAACTATACCGGTGGGGATTTAGATGCCAATATGCAAAAAGCCGCTTGGCGCTTGGGCCAATCCAATGGCAGCTCTTCCACAAACTTTAAAGAGGCTGCCAATCGCGCCACAAGGGTGAATTCCAACGCTAAAAATGTCATAATCGGTAATTTTTTAGAGATCCATAACCGCATGGGTTCTGGGGCTAGGAGAAAAGCCAGCTCTGCGATTTTAACTTTACAAAGTGCAAGAAGGGATTGCAAGCGCTGAAAACGCTGAAATTTCTCTTTATGATGGCACTACCCTTAATTTGCAAGGAGCGCGCGCTGTGAATTTATTGGGTAAGGTGTGGATGGGCCGTTTGCAATATGCGGGAGCGTATTTAGCCCCTTCATGGAGCATGGTGAAGCATCTTGGCATTTTAAATACCCACTACAATATTTCGCATTTGTTCTAGCGCCCGATAAACACTTCAAAGGTGCAAGGGGATGTGAATTGGTGCCATCTCGTTGTGGGTGATCACAACGCCGCTCAAGCAGGCATTGCTGCTAATGAAAAGACTGACATTGCGCGCTGGATTTGTGACAAAGCATAGGATTAAACCTTATTGCTTCTCCAGAAGGTGGCTATAAAAAGACCACTTCTCAAAATAACGCTGACAATACCCAACAAAATAGCAGCCAAACGAATAGCCAAAACCATGAATGGCGCTCAAGCCATTAACCCGCCCAATAGCGTGCAAAAAGTAGAAGTTGAACCCATGAAAGTCATTGACGGGACTTTTGCTGGCGCTAGGGGCACGGTCGTGAATATTTTCCACTGAAACACTGACGCTGATGGGACGATTAAAGCGGGAGGGCATAAAGCTTTTGTTAGCGCCAGTGCGGCTCATTTGAATGTCGGCGAAGGCGCTATCCATCTATCCAATTCAACGAGTGGGCGCGCTCTTTTAGCGGGAAATCTAACCGGGAATACCACCGTTAATGGGGCATTAAGGGTGAATAAAGAAGTGGGTGGTGCTTTGCCAGGTTCAAGCGCGAATTTTGAATTTAAAGTTGGTGTGGATACCCAAAAGGGTGCAGCCGCTTTTAATAACGATACCCGTTTGGGAAAAGCGGTGGATTTAAAAGCGGGTGTGCATGCAGTTAATTTTAATGGTGATGCGTATTGAGGAAAATTTACGCATTTTAAAGGGGGTGCCAATACCGCTAGTTTTAAAGATATTGATGCCAATAAGGGTGGTAATGGCATCAACGCTACCACTTTGGATTTTAGGAACGTTATTAGGCATGGCCAATACCAACAAGCACAGCTTCCACTCATGCTGCCATTGAAAACTTCAGTATTAAGGAATTGGTTGTTACAACCCATGGTATGAGTGCGGGAAAATGCTCTCCATTTAGCGGGGATATAGGCGATAAATCGCGCATTACTACCGTGCGTTTAGAAACGGGATATAGCCCGCTCTATGCTAGAGGGGTTAAATTTAAAAGCGGTAAAAAACCCGTTATAGATGGGATGTGCCATGCTCCTTGGAATTATTTTGACGCCAGAAACACTACCGGTGTTGAAACCACCAAGAAAATTCTTTTTGGAGCCCTAGGAAACCTTGCAGGCAAGATAGGGCTTAATAACCTAACCTTGGGTCATCATGCAAGCATGGATTATGGTAAAGATTTGGATTTAACCCTTCAGGGGCATTTTATTAACAATCAAGGCGTGATGGATCTTTTTGTTCAAAACTGGCGTGCAGCAACCTTGAATGCAAGCCATCAAGCAAGCATGCTGTTTAATAACATGACAGACAACACAACCGGATTTTACAAACCGCTCATTAAGATCAATAACGCTCAAAACCTCACTAAAAGCACAGAGCATGTTTTTGGTGGGAGCGCGAAACATTGATGATCATTGAGCGGGAGTGCAAGGCGTTAGTTATGATGCTGTTTTTGCAAGCAATGCTAATCTGCAAGAGCAATTCAAAGAGTGCCTAGCCCTTTATAACAACAATACCCGTATGGGTATTTGTGTGGCGCAAAAAGGCAACACCAACGGTATTAAAGCATGCGGAATGGCTATCGGTAATCATGACATGAACAGTAACCCTAACCATTACAAGTATCTTGAAGGTAAGGCATGGAAAAATACAGGCATTAGTCAAACAGCCAACAACGCTAAAACCTCTGTTCGTTTGGGTAACAATTCTATGCCTAGTGGGCATGACAACAATGCTGCAAATTGACCCACAAACAACAGACAGGCGCGTTTCCTACGCTCTCATAAAAAACGCTCCTTTGGTCAAACGCAAGCGCCTCTTCTAAAGCGCCACTCCTAATTTAGTCCACATCCATGAGCCTCATTTTGGCACCCTTGAGAGCGCGTTTGAATTGGCTAACCGCTCCAAAGATATTGACGCGCTTTATGCCCATTCAGGCACGCAAGGCAGGAATCTCTTGCAAACCTTATTGGTTGATCATCATGATATTAGGCTATGCCAGAGAAATGATTGATAACACAAGTGCCAGCGAAACCACCAAGCAATTGGATGCGGCAACTGACACTTGAAACAGCATAGCCCGTTTGGATCGCAAAACAAGTGGTTTGCAAACTTTGAGTTTGAGCAATGCGATGGTTTTGAATTCTCGTTTAGCCAATCTCTCCAGGAAGCGCACCAACAATATTGGCTCATTCGCTAAACGCTTGCAAGCTTTAAAAGACCAAAAATTTAGCCCCACAGAAAGCGCGGCAGAAGTGTTGTATAAGTTCGCTCCTAACCATGAAAAACTTAAAAATCTTTGGGCTAATGCTATCAAGGGAGCGAGTTTGAATAGTGGTGGCAACACTTCATTATAGGGCACAAGCGCAGGGGTGGATACTTGCATTAACGGGGAAGTGGAAGCTCTTGTTGACGGGTTTGGAAGCGATGGTTATAGCGCCTTCAGTAACCGCGCTAGCCCTCTTAACTCTGATGGCAATAACGCTCATTTTGGCGTGTATAACCGTATCTTGGCTGACCAACAAAAATTTGACTTTCAAGCTGGGCTGTGGAAAGCGATCAATCAAGCTTGAATTGCAAAGGCGCTTTGTGGCGCTATGGCTAACGCCAGCTATGGTTATGGCTTTGGGTTTTTCAAAAACGCCTTAGCGCTAAAACCAAGCGTGGGAGTGATTCATAATCATTTAGGTTCAACCAACTTTGAAAGCAATAGCGCCCAAAAGCGGCTTTAAGCGATGGTGCAAGCGGTCAGCATTTATCCAACGCTGACGGCAACCTACAAGCCCGCTATTATTGCCGAGACACTTCATGCTTCTGTTTGCATGCTGGAGTTTTACAAGAGTTTGCCAACTTTGAATCCAACCACGCTGTGTCTTTGAACGCCTTCAAAGCAAATGCCGCTCGCAACCCTTTAAACACGCATGCCAGAGTGATGATGGGTGGGGAATTGGATTTGGCTAAAGAAGTGTTTTTGAATTGGGGCTTCATTGATTTGGACAACTTGATTTCCAATGCGAGCCGTTTCGCTTCCAGTTG\n",
      "MEPQQTHHKHRKIKRPMVPLA\n",
      "1317 21\n"
     ]
    }
   ],
   "source": [
    "file = GBFFFile('../data/proteins/ncbi/GCF_000009305.1_genomic.gbff')\n",
    "file.translate(459)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "top_hit_genome_id GCF_000259275.1 391\n",
      "query_strand -1 top_hit_strand -1\n",
      "query_seq (402789, 403010) MDFGQKWLGVSVSQNYKNAIKPLFETLEKAKKEGMLWRNFSNKEQEIYAPLLQAFKDEVLRIDENKKNKVPQK\n",
      "top_hit_seq (402451, 403402, codon_start=1) LSAKSNNQGRAYEFAWLLVLEQKLSVFKKVIVDKQGGYNASCRAYESLEKSLQEHYLASAKSGVLLLLDCEPLLSEILDSSHNEIKLSLQTDKLGEIADVRDILIHFDTFCIGLSINIIMRL\n",
      "overlap_length (402789, 403010) 222\n",
      "query_overlap_fraction 1.0 top_hit_overlap_fraction 0.2331932773109243\n",
      "top_hit_product HaeIII family restriction endonuclease\n",
      "n_hits 1\n",
      "top_hit_note frameshifted; Derived by automated computational analysis using gene prediction method: Protein Homology.\n"
     ]
    }
   ],
   "source": [
    "def info(ref_df:pd.DataFrame, id_:str):\n",
    "    row = ref_df.loc[id_]\n",
    "    print('top_hit_genome_id', row.top_hit_genome_id, f'{row.top_hit_id}')\n",
    "    print('query_strand', row.query_strand, 'top_hit_strand', row.top_hit_strand)\n",
    "    print('query_seq', f'({row.query_start}, {row.query_stop})', row.query_seq)\n",
    "    print('top_hit_seq', f'({row.top_hit_start}, {row.top_hit_stop}, codon_start={row.top_hit_codon_start})', row.top_hit_seq)\n",
    "    print('overlap_length', f'({row.overlap_start}, {row.overlap_stop})', row.overlap_length)\n",
    "    print('query_overlap_fraction', row.query_overlap_fraction, 'top_hit_overlap_fraction', row.top_hit_overlap_fraction)\n",
    "    print('top_hit_product', row.top_hit_product)\n",
    "    print('n_hits', row.n_hits)\n",
    "    print('top_hit_note', row.top_hit_note)\n",
    "\n",
    "\n",
    "info(ref_df, 'NC_017735.1_382')\n",
    "# 443855"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>top_hit_product</th>\n",
       "      <th>query_overlap_fraction</th>\n",
       "      <th>top_hit_overlap_fraction</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>same_strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_452</th>\n",
       "      <td>vacuolating cyotoxin family protein</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.056933</td>\n",
       "      <td>225</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     top_hit_product  query_overlap_fraction  \\\n",
       "id                                                                             \n",
       "NC_008229.1_452  vacuolating cyotoxin family protein                     1.0   \n",
       "\n",
       "                 top_hit_overlap_fraction  overlap_length  same_strand  \n",
       "id                                                                      \n",
       "NC_008229.1_452                  0.056933             225         True  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_spurious_df[ref_spurious_df.index == 'NC_008229.1_452'][['top_hit_product', 'query_overlap_fraction', 'top_hit_overlap_fraction', 'overlap_length', 'same_strand']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>analysis</th>\n",
       "      <th>signature</th>\n",
       "      <th>signature_description</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>e_value</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>annotation</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NC_004917.1_939</th>\n",
       "      <td>573</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF11186</td>\n",
       "      <td>Protein of unknown function (DUF2972)</td>\n",
       "      <td>369</td>\n",
       "      <td>560</td>\n",
       "      <td>1.600000e-74</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR021353</td>\n",
       "      <td>Protein of unknown function DUF2972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_1094</th>\n",
       "      <td>96</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02691</td>\n",
       "      <td>Vacuolating cyotoxin</td>\n",
       "      <td>1</td>\n",
       "      <td>91</td>\n",
       "      <td>7.800000e-37</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003842</td>\n",
       "      <td>Vacuolating cytotoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_1098</th>\n",
       "      <td>176</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02691</td>\n",
       "      <td>Vacuolating cyotoxin</td>\n",
       "      <td>2</td>\n",
       "      <td>170</td>\n",
       "      <td>4.800000e-71</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003842</td>\n",
       "      <td>Vacuolating cytotoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_1100</th>\n",
       "      <td>74</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02691</td>\n",
       "      <td>Vacuolating cyotoxin</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>9.300000e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003842</td>\n",
       "      <td>Vacuolating cytotoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_1472</th>\n",
       "      <td>113</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01856</td>\n",
       "      <td>Helicobacter outer membrane protein</td>\n",
       "      <td>5</td>\n",
       "      <td>108</td>\n",
       "      <td>2.800000e-15</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR002718</td>\n",
       "      <td>Outer membrane protein, Helicobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_452</th>\n",
       "      <td>74</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02691</td>\n",
       "      <td>Vacuolating cyotoxin</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>9.300000e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003842</td>\n",
       "      <td>Vacuolating cytotoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_454</th>\n",
       "      <td>176</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02691</td>\n",
       "      <td>Vacuolating cyotoxin</td>\n",
       "      <td>2</td>\n",
       "      <td>170</td>\n",
       "      <td>4.800000e-71</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003842</td>\n",
       "      <td>Vacuolating cytotoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_458</th>\n",
       "      <td>96</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02691</td>\n",
       "      <td>Vacuolating cyotoxin</td>\n",
       "      <td>1</td>\n",
       "      <td>91</td>\n",
       "      <td>7.800000e-37</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003842</td>\n",
       "      <td>Vacuolating cytotoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_542</th>\n",
       "      <td>118</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF07510</td>\n",
       "      <td>GmrSD restriction endonucleases, C-terminal do...</td>\n",
       "      <td>46</td>\n",
       "      <td>117</td>\n",
       "      <td>4.300000e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR011089</td>\n",
       "      <td>GmrSD restriction endonucleases, C-terminal do...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_008229.1_716</th>\n",
       "      <td>125</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF09522</td>\n",
       "      <td>R.Pab1 restriction endonuclease</td>\n",
       "      <td>37</td>\n",
       "      <td>98</td>\n",
       "      <td>1.600000e-27</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR018576</td>\n",
       "      <td>Restriction endonuclease, type II, Pab1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_017735.1_382</th>\n",
       "      <td>73</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF09556</td>\n",
       "      <td>HaeIII restriction endonuclease</td>\n",
       "      <td>1</td>\n",
       "      <td>73</td>\n",
       "      <td>4.800000e-18</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR019059</td>\n",
       "      <td>Restriction endonuclease, type II, HaeIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP022826.1_324</th>\n",
       "      <td>86</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF12762</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>11</td>\n",
       "      <td>86</td>\n",
       "      <td>4.300000e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR024445</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AP023046.1_1650</th>\n",
       "      <td>203</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02534</td>\n",
       "      <td>Type IV secretory system Conjugative DNA transfer</td>\n",
       "      <td>91</td>\n",
       "      <td>201</td>\n",
       "      <td>6.000000e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003688</td>\n",
       "      <td>Type IV secretion system protein TraG/VirD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CDMK01000001.1_633</th>\n",
       "      <td>112</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01131</td>\n",
       "      <td>DNA topoisomerase</td>\n",
       "      <td>3</td>\n",
       "      <td>105</td>\n",
       "      <td>2.300000e-13</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR013497</td>\n",
       "      <td>DNA topoisomerase, type IA, central</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP007770.1_215</th>\n",
       "      <td>58</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF05787</td>\n",
       "      <td>Alkaline phosphatase PhoX</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "      <td>4.000000e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR008557</td>\n",
       "      <td>Alkaline phosphatase PhoX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP012543.1_2150</th>\n",
       "      <td>93</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF12762</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>3</td>\n",
       "      <td>81</td>\n",
       "      <td>2.800000e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR024445</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP031611.1_1252</th>\n",
       "      <td>100</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF17236</td>\n",
       "      <td>SU10 major capsid protein</td>\n",
       "      <td>9</td>\n",
       "      <td>96</td>\n",
       "      <td>1.600000e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR035198</td>\n",
       "      <td>SU10 major capsid protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP031611.1_1369</th>\n",
       "      <td>131</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF14287</td>\n",
       "      <td>Domain of unknown function (DUF4368)</td>\n",
       "      <td>74</td>\n",
       "      <td>125</td>\n",
       "      <td>4.000000e-12</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR025378</td>\n",
       "      <td>Domain of unknown function DUF4368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP043427.1_1636</th>\n",
       "      <td>114</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF05935</td>\n",
       "      <td>Arylsulfotransferase (ASST)</td>\n",
       "      <td>19</td>\n",
       "      <td>107</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR010262</td>\n",
       "      <td>Arylsulfotransferase, bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP145316.1_909</th>\n",
       "      <td>82</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02086</td>\n",
       "      <td>D12 class N6 adenine-specific DNA methyltransf...</td>\n",
       "      <td>9</td>\n",
       "      <td>78</td>\n",
       "      <td>5.000000e-14</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR012327</td>\n",
       "      <td>D12 class N6 adenine-specific DNA methyltransf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_FZMA01000017.1_2</th>\n",
       "      <td>104</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF07282</td>\n",
       "      <td>Cas12f1-like, TNB domain</td>\n",
       "      <td>26</td>\n",
       "      <td>91</td>\n",
       "      <td>7.300000e-17</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR010095</td>\n",
       "      <td>Cas12f1-like, TNB domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_JAJTNB010000012.1_24</th>\n",
       "      <td>113</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02537</td>\n",
       "      <td>CrcB-like protein, Camphor Resistance (CrcB)</td>\n",
       "      <td>13</td>\n",
       "      <td>67</td>\n",
       "      <td>1.500000e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR003691</td>\n",
       "      <td>Fluoride-specific ion channel FluC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_JANURX010000003.1_179</th>\n",
       "      <td>47</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01964</td>\n",
       "      <td>Radical SAM ThiC family</td>\n",
       "      <td>2</td>\n",
       "      <td>46</td>\n",
       "      <td>3.900000e-15</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR002817</td>\n",
       "      <td>Phosphomethylpyrimidine synthase ThiC/5-hydrox...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_NXLZ01000010.1_18</th>\n",
       "      <td>227</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01964</td>\n",
       "      <td>Radical SAM ThiC family</td>\n",
       "      <td>4</td>\n",
       "      <td>227</td>\n",
       "      <td>1.700000e-80</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR002817</td>\n",
       "      <td>Phosphomethylpyrimidine synthase ThiC/5-hydrox...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_NXLZ01000010.1_19</th>\n",
       "      <td>153</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF04301</td>\n",
       "      <td>Pimeloyl-ACP methyl esterase BioG</td>\n",
       "      <td>2</td>\n",
       "      <td>143</td>\n",
       "      <td>3.900000e-15</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR007398</td>\n",
       "      <td>Pimeloyl-ACP methyl esterase BioG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_NXLZ01000010.1_20</th>\n",
       "      <td>58</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01964</td>\n",
       "      <td>Radical SAM ThiC family</td>\n",
       "      <td>1</td>\n",
       "      <td>52</td>\n",
       "      <td>9.700000e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR002817</td>\n",
       "      <td>Phosphomethylpyrimidine synthase ThiC/5-hydrox...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QHLI01000008.1_31</th>\n",
       "      <td>85</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF12762</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>2</td>\n",
       "      <td>67</td>\n",
       "      <td>3.700000e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR024445</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QHLI01000014.1_37</th>\n",
       "      <td>109</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF12762</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>5</td>\n",
       "      <td>86</td>\n",
       "      <td>1.100000e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR024445</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_QXQQ01000026.1_31</th>\n",
       "      <td>107</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01385</td>\n",
       "      <td>Probable transposase</td>\n",
       "      <td>17</td>\n",
       "      <td>96</td>\n",
       "      <td>1.200000e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>04-03-2025</td>\n",
       "      <td>IPR001959</td>\n",
       "      <td>Probable transposase, IS891/IS1136/IS1341</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          length analysis signature  \\\n",
       "id                                                    \n",
       "NC_004917.1_939              573     Pfam   PF11186   \n",
       "NC_008229.1_1094              96     Pfam   PF02691   \n",
       "NC_008229.1_1098             176     Pfam   PF02691   \n",
       "NC_008229.1_1100              74     Pfam   PF02691   \n",
       "NC_008229.1_1472             113     Pfam   PF01856   \n",
       "NC_008229.1_452               74     Pfam   PF02691   \n",
       "NC_008229.1_454              176     Pfam   PF02691   \n",
       "NC_008229.1_458               96     Pfam   PF02691   \n",
       "NC_008229.1_542              118     Pfam   PF07510   \n",
       "NC_008229.1_716              125     Pfam   PF09522   \n",
       "NC_017735.1_382               73     Pfam   PF09556   \n",
       "NZ_AP022826.1_324             86     Pfam   PF12762   \n",
       "NZ_AP023046.1_1650           203     Pfam   PF02534   \n",
       "NZ_CDMK01000001.1_633        112     Pfam   PF01131   \n",
       "NZ_CP007770.1_215             58     Pfam   PF05787   \n",
       "NZ_CP012543.1_2150            93     Pfam   PF12762   \n",
       "NZ_CP031611.1_1252           100     Pfam   PF17236   \n",
       "NZ_CP031611.1_1369           131     Pfam   PF14287   \n",
       "NZ_CP043427.1_1636           114     Pfam   PF05935   \n",
       "NZ_CP145316.1_909             82     Pfam   PF02086   \n",
       "NZ_FZMA01000017.1_2          104     Pfam   PF07282   \n",
       "NZ_JAJTNB010000012.1_24      113     Pfam   PF02537   \n",
       "NZ_JANURX010000003.1_179      47     Pfam   PF01964   \n",
       "NZ_NXLZ01000010.1_18         227     Pfam   PF01964   \n",
       "NZ_NXLZ01000010.1_19         153     Pfam   PF04301   \n",
       "NZ_NXLZ01000010.1_20          58     Pfam   PF01964   \n",
       "NZ_QHLI01000008.1_31          85     Pfam   PF12762   \n",
       "NZ_QHLI01000014.1_37         109     Pfam   PF12762   \n",
       "NZ_QXQQ01000026.1_31         107     Pfam   PF01385   \n",
       "\n",
       "                                                      signature_description  \\\n",
       "id                                                                            \n",
       "NC_004917.1_939                       Protein of unknown function (DUF2972)   \n",
       "NC_008229.1_1094                                       Vacuolating cyotoxin   \n",
       "NC_008229.1_1098                                       Vacuolating cyotoxin   \n",
       "NC_008229.1_1100                                       Vacuolating cyotoxin   \n",
       "NC_008229.1_1472                        Helicobacter outer membrane protein   \n",
       "NC_008229.1_452                                        Vacuolating cyotoxin   \n",
       "NC_008229.1_454                                        Vacuolating cyotoxin   \n",
       "NC_008229.1_458                                        Vacuolating cyotoxin   \n",
       "NC_008229.1_542           GmrSD restriction endonucleases, C-terminal do...   \n",
       "NC_008229.1_716                             R.Pab1 restriction endonuclease   \n",
       "NC_017735.1_382                             HaeIII restriction endonuclease   \n",
       "NZ_AP022826.1_324                             ISXO2-like transposase domain   \n",
       "NZ_AP023046.1_1650        Type IV secretory system Conjugative DNA transfer   \n",
       "NZ_CDMK01000001.1_633                                     DNA topoisomerase   \n",
       "NZ_CP007770.1_215                                 Alkaline phosphatase PhoX   \n",
       "NZ_CP012543.1_2150                            ISXO2-like transposase domain   \n",
       "NZ_CP031611.1_1252                                SU10 major capsid protein   \n",
       "NZ_CP031611.1_1369                     Domain of unknown function (DUF4368)   \n",
       "NZ_CP043427.1_1636                              Arylsulfotransferase (ASST)   \n",
       "NZ_CP145316.1_909         D12 class N6 adenine-specific DNA methyltransf...   \n",
       "NZ_FZMA01000017.1_2                                Cas12f1-like, TNB domain   \n",
       "NZ_JAJTNB010000012.1_24        CrcB-like protein, Camphor Resistance (CrcB)   \n",
       "NZ_JANURX010000003.1_179                            Radical SAM ThiC family   \n",
       "NZ_NXLZ01000010.1_18                                Radical SAM ThiC family   \n",
       "NZ_NXLZ01000010.1_19                      Pimeloyl-ACP methyl esterase BioG   \n",
       "NZ_NXLZ01000010.1_20                                Radical SAM ThiC family   \n",
       "NZ_QHLI01000008.1_31                          ISXO2-like transposase domain   \n",
       "NZ_QHLI01000014.1_37                          ISXO2-like transposase domain   \n",
       "NZ_QXQQ01000026.1_31                                   Probable transposase   \n",
       "\n",
       "                          start  stop       e_value  status        date  \\\n",
       "id                                                                        \n",
       "NC_004917.1_939             369   560  1.600000e-74    True  04-03-2025   \n",
       "NC_008229.1_1094              1    91  7.800000e-37    True  04-03-2025   \n",
       "NC_008229.1_1098              2   170  4.800000e-71    True  04-03-2025   \n",
       "NC_008229.1_1100              1    33  9.300000e-09    True  04-03-2025   \n",
       "NC_008229.1_1472              5   108  2.800000e-15    True  04-03-2025   \n",
       "NC_008229.1_452               1    33  9.300000e-09    True  04-03-2025   \n",
       "NC_008229.1_454               2   170  4.800000e-71    True  04-03-2025   \n",
       "NC_008229.1_458               1    91  7.800000e-37    True  04-03-2025   \n",
       "NC_008229.1_542              46   117  4.300000e-06    True  04-03-2025   \n",
       "NC_008229.1_716              37    98  1.600000e-27    True  04-03-2025   \n",
       "NC_017735.1_382               1    73  4.800000e-18    True  04-03-2025   \n",
       "NZ_AP022826.1_324            11    86  4.300000e-11    True  04-03-2025   \n",
       "NZ_AP023046.1_1650           91   201  6.000000e-16    True  04-03-2025   \n",
       "NZ_CDMK01000001.1_633         3   105  2.300000e-13    True  04-03-2025   \n",
       "NZ_CP007770.1_215             1    51  4.000000e-11    True  04-03-2025   \n",
       "NZ_CP012543.1_2150            3    81  2.800000e-06    True  04-03-2025   \n",
       "NZ_CP031611.1_1252            9    96  1.600000e-11    True  04-03-2025   \n",
       "NZ_CP031611.1_1369           74   125  4.000000e-12    True  04-03-2025   \n",
       "NZ_CP043427.1_1636           19   107  1.000000e-07    True  04-03-2025   \n",
       "NZ_CP145316.1_909             9    78  5.000000e-14    True  04-03-2025   \n",
       "NZ_FZMA01000017.1_2          26    91  7.300000e-17    True  04-03-2025   \n",
       "NZ_JAJTNB010000012.1_24      13    67  1.500000e-10    True  04-03-2025   \n",
       "NZ_JANURX010000003.1_179      2    46  3.900000e-15    True  04-03-2025   \n",
       "NZ_NXLZ01000010.1_18          4   227  1.700000e-80    True  04-03-2025   \n",
       "NZ_NXLZ01000010.1_19          2   143  3.900000e-15    True  04-03-2025   \n",
       "NZ_NXLZ01000010.1_20          1    52  9.700000e-10    True  04-03-2025   \n",
       "NZ_QHLI01000008.1_31          2    67  3.700000e-06    True  04-03-2025   \n",
       "NZ_QHLI01000014.1_37          5    86  1.100000e-08    True  04-03-2025   \n",
       "NZ_QXQQ01000026.1_31         17    96  1.200000e-08    True  04-03-2025   \n",
       "\n",
       "                         annotation  \\\n",
       "id                                    \n",
       "NC_004917.1_939           IPR021353   \n",
       "NC_008229.1_1094          IPR003842   \n",
       "NC_008229.1_1098          IPR003842   \n",
       "NC_008229.1_1100          IPR003842   \n",
       "NC_008229.1_1472          IPR002718   \n",
       "NC_008229.1_452           IPR003842   \n",
       "NC_008229.1_454           IPR003842   \n",
       "NC_008229.1_458           IPR003842   \n",
       "NC_008229.1_542           IPR011089   \n",
       "NC_008229.1_716           IPR018576   \n",
       "NC_017735.1_382           IPR019059   \n",
       "NZ_AP022826.1_324         IPR024445   \n",
       "NZ_AP023046.1_1650        IPR003688   \n",
       "NZ_CDMK01000001.1_633     IPR013497   \n",
       "NZ_CP007770.1_215         IPR008557   \n",
       "NZ_CP012543.1_2150        IPR024445   \n",
       "NZ_CP031611.1_1252        IPR035198   \n",
       "NZ_CP031611.1_1369        IPR025378   \n",
       "NZ_CP043427.1_1636        IPR010262   \n",
       "NZ_CP145316.1_909         IPR012327   \n",
       "NZ_FZMA01000017.1_2       IPR010095   \n",
       "NZ_JAJTNB010000012.1_24   IPR003691   \n",
       "NZ_JANURX010000003.1_179  IPR002817   \n",
       "NZ_NXLZ01000010.1_18      IPR002817   \n",
       "NZ_NXLZ01000010.1_19      IPR007398   \n",
       "NZ_NXLZ01000010.1_20      IPR002817   \n",
       "NZ_QHLI01000008.1_31      IPR024445   \n",
       "NZ_QHLI01000014.1_37      IPR024445   \n",
       "NZ_QXQQ01000026.1_31      IPR001959   \n",
       "\n",
       "                                                                description  \n",
       "id                                                                           \n",
       "NC_004917.1_939                         Protein of unknown function DUF2972  \n",
       "NC_008229.1_1094                                      Vacuolating cytotoxin  \n",
       "NC_008229.1_1098                                      Vacuolating cytotoxin  \n",
       "NC_008229.1_1100                                      Vacuolating cytotoxin  \n",
       "NC_008229.1_1472                       Outer membrane protein, Helicobacter  \n",
       "NC_008229.1_452                                       Vacuolating cytotoxin  \n",
       "NC_008229.1_454                                       Vacuolating cytotoxin  \n",
       "NC_008229.1_458                                       Vacuolating cytotoxin  \n",
       "NC_008229.1_542           GmrSD restriction endonucleases, C-terminal do...  \n",
       "NC_008229.1_716                     Restriction endonuclease, type II, Pab1  \n",
       "NC_017735.1_382                   Restriction endonuclease, type II, HaeIII  \n",
       "NZ_AP022826.1_324                             ISXO2-like transposase domain  \n",
       "NZ_AP023046.1_1650              Type IV secretion system protein TraG/VirD4  \n",
       "NZ_CDMK01000001.1_633                   DNA topoisomerase, type IA, central  \n",
       "NZ_CP007770.1_215                                 Alkaline phosphatase PhoX  \n",
       "NZ_CP012543.1_2150                            ISXO2-like transposase domain  \n",
       "NZ_CP031611.1_1252                                SU10 major capsid protein  \n",
       "NZ_CP031611.1_1369                       Domain of unknown function DUF4368  \n",
       "NZ_CP043427.1_1636                           Arylsulfotransferase, bacteria  \n",
       "NZ_CP145316.1_909         D12 class N6 adenine-specific DNA methyltransf...  \n",
       "NZ_FZMA01000017.1_2                                Cas12f1-like, TNB domain  \n",
       "NZ_JAJTNB010000012.1_24                  Fluoride-specific ion channel FluC  \n",
       "NZ_JANURX010000003.1_179  Phosphomethylpyrimidine synthase ThiC/5-hydrox...  \n",
       "NZ_NXLZ01000010.1_18      Phosphomethylpyrimidine synthase ThiC/5-hydrox...  \n",
       "NZ_NXLZ01000010.1_19                      Pimeloyl-ACP methyl esterase BioG  \n",
       "NZ_NXLZ01000010.1_20      Phosphomethylpyrimidine synthase ThiC/5-hydrox...  \n",
       "NZ_QHLI01000008.1_31                          ISXO2-like transposase domain  \n",
       "NZ_QHLI01000014.1_37                          ISXO2-like transposase domain  \n",
       "NZ_QXQQ01000026.1_31              Probable transposase, IS891/IS1136/IS1341  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = InterProScanFile('../data/campylobacterota_spurious_interpro.tsv')\n",
    "df = file.to_df(max_e_value=1e-5)\n",
    "df = df.sort_values('e_value').groupby(df.index).first()\n",
    "df[df.analysis == 'Pfam']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "analysis\n",
       "AntiFam        294\n",
       "Gene3D          46\n",
       "PANTHER         41\n",
       "Pfam            29\n",
       "SUPERFAMILY     28\n",
       "NCBIfam         14\n",
       "CDD              1\n",
       "SMART            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = InterProScanFile('../data/campylobacterota_intergenic_interpro.tsv')\n",
    "df = file.to_df(max_e_value=1e-5)\n",
    "df = df.sort_values('e_value').groupby(df.index).first()\n",
    "df.value_counts('analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def check_spurious(ref_df:pd.DataFrame, max_score:float=0.5, remove_invalid:bool=True):\n",
    "    scores = get_alignment_scores(ref_df, seq_a_col='top_hit_seq', seq_b_col='query_seq')\n",
    "    # ref_df['alignment_score'] = scores\n",
    "\n",
    "    valid = ref_df.top_hit_feature.isin(GBFFFile.rna_features + GBFFFile.other_features)\n",
    "    valid = valid | ((scores < max_score) & ~np.isnan(scores)) \n",
    "    ref_df['valid'] = valid\n",
    "\n",
    "    print(f'check_spurious: Validated {ref_df.valid.sum()} out of {len(ref_df)} spurious sequences.')\n",
    "\n",
    "    if remove_invalid:\n",
    "        ref_df = ref_df[ref_df.valid].copy()\n",
    "        ref_df = ref_df.drop(columns=['valid'])\n",
    "    return ref_df\n",
    "    \n",
    "\n",
    "def check_real(ref_df:pd.DataFrame, min_score:float=0.9, remove_invalid:bool=True):\n",
    "    scores = get_alignment_scores(ref_df, seq_a_col='top_hit_seq', seq_b_col='query_seq')\n",
    "    # ref_df['alignment_score'] = scores\n",
    "    ref_df['valid'] = (scores > min_score) & ~np.isnan(scores)\n",
    "\n",
    "    print(f'check_real: Validated {ref_df.valid.sum()} out of {len(ref_df)} real sequences.')\n",
    "    \n",
    "    if remove_invalid:\n",
    "        ref_df = ref_df[ref_df.valid].copy()\n",
    "        ref_df = ref_df.drop(columns=['valid'])\n",
    "    return ref_df\n",
    "\n",
    "# It seems that only including things with definitive labels seems to have just increased the number of false negatives. \n",
    "# Possibly the benefits of increasing label confidence were outweighed by the shrinkage of the minority class?\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "build_datasets: Loading data from 196 genomes belonging to the phylum Campylobacterota.\n",
      "build_datasets: Loaded 409652 sequences.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "get_alignment_scores: 100%|| 362090/362090 [11:32<00:00, 522.68it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "check_real: Validated 360736 out of 362090 real sequences.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "get_alignment_scores: 100%|| 1015/1015 [00:00<00:00, 7047.43it/s]\n",
      "/home/prichter/Documents/tripy/notebooks/utils.py:72: UserWarning: get_lengths: Not all gene lengths are divisible by three.\n",
      "  warnings.warn('get_lengths: Not all gene lengths are divisible by three.')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "check_spurious: Validated 1014 out of 1015 spurious sequences.\n",
      "build_datasets: Kept 1014 spurious sequences and 360736 real sequences.\n",
      "Removing 345 sequences exceeding the maximum length of 2000\n",
      "MMseqs.load: Removing 302432 non-cluster representatives.\n",
      "build_datasets: Clustering at 50 percent similarity removed 357 spurious sequences and 302075 sequences.\n",
      "build_datasets: Training dataset contains 45811 sequences, testing dataset contains 13162 sequences.\n",
      "build_datasets: 541 negative instances and 45270 positive instances in the training dataset.\n",
      "build_datasets: 116 negative instances and 13046 positive instances in the testing dataset.\n"
     ]
    }
   ],
   "source": [
    "def build_datasets(genome_metadata_df:pd.DataFrame, phylum:str='Campylobacterota', max_length:int=2000):\n",
    "    # First remove things which do not have definitive labels. \n",
    "    genome_ids = genome_metadata_df[genome_metadata_df.phylum == phylum].index\n",
    "    genome_ids = [genome_id for genome_id in genome_ids if (genome_id != 'GCF_000009085.1')] # Remove the non-PGAP genome. \n",
    "\n",
    "    print(f'build_datasets: Loading data from {len(genome_ids)} genomes belonging to the phylum {phylum}.')\n",
    "    df = load_ref(genome_ids=genome_ids)\n",
    "    print(f'build_datasets: Loaded {len(ref_df)} sequences.')\n",
    "\n",
    "    df = pd.concat([check_real(df[df.real].copy()), check_spurious(df[df.spurious].copy())])\n",
    "    print(f'build_datasets: Kept {df.spurious.sum()} spurious sequences and {df.real.sum()} real sequences.')\n",
    "    df['label'] = df.real.astype(int)\n",
    "\n",
    "    lengths = get_lengths(df, top_hit=False)\n",
    "    print(f'Removing {(lengths >= max_length).sum()} sequences exceeding the maximum length of {max_length}')\n",
    "    df = df[lengths < max_length]\n",
    "\n",
    "    df = df.rename(columns={'query_seq':'seq'}) # Need to do this for file writing, etc. to work correctly, \n",
    "    df = df.rename(columns={'top_hit_genome_id':'genome_id'}) # Need to do this for file writing, etc. to work correctly, \n",
    "\n",
    "    mmseqs = MMseqs()\n",
    "    n_real, n_spurious = df.real.sum(), df.spurious.sum()\n",
    "    df = mmseqs.cluster(df, job_name=f'{phylum.lower()}_50', sequence_identity=0.50, reps_only=True, overwrite=True)\n",
    "    print(f'build_datasets: Clustering at 50 percent similarity removed {n_spurious - df.spurious.sum()} spurious sequences and {n_real - df.real.sum()} sequences.')\n",
    "    mmseqs.cleanup()\n",
    "\n",
    "    gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "    train_idxs, test_idxs = list(gss.split(df, groups=df.genome_id))[0]\n",
    "    train_df, test_df = df.iloc[train_idxs], df.iloc[test_idxs]\n",
    "    print(f'build_datasets: Training dataset contains {len(train_df)} sequences, testing dataset contains {len(test_df)} sequences.')\n",
    "    print(f'build_datasets: {(train_df.label == 0).sum()} negative instances and {(train_df.label == 1).sum()} positive instances in the training dataset.')\n",
    "    print(f'build_datasets: {(test_df.label == 0).sum()} negative instances and {(test_df.label == 1).sum()} positive instances in the testing dataset.')\n",
    "    train_df.to_csv(f'../data/{phylum.lower()}_dataset_train.csv')\n",
    "    test_df.to_csv(f'../data/{phylum.lower()}_dataset_test.csv')\n",
    "    return train_df, test_df \n",
    "    \n",
    "train_df, test_df = build_datasets(genome_metadata_df)\n",
    "phylum ='Campylobacterota'\n",
    "# dtypes = {'top_hit_partial':str, 'query_partial':str, 'top_hit_translation_table':str, 'top_hit_codon_start':str}\n",
    "# train_df = pd.read_csv(f'../data/{phylum.lower()}_dataset_train.csv', index_col=0, dtype=dtypes)\n",
    "# test_df = pd.read_csv(f'../data/{phylum.lower()}_dataset_test.csv', index_col=0, dtype=dtypes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred_df = load_predict('../data/predict/campylobacterota_dataset_test_predict.csv', model_name='campylobacterota_esm_650m_gap_v1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Balanced accuracy: 0.8243785650516682\n",
      "Precision (class 0): 0.0003928135789451922\n",
      "Recall (class 0): 0.12666666666666668\n",
      "Number of false negatives: 14003\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['NC_017735.1_1005', 'NC_017735.1_1006', 'NC_017735.1_1007', ...,\n",
       "       'NZ_WFKI01000156.1_1', 'NZ_WFKI01000159.1_1',\n",
       "       'NZ_WFKI01000162.1_1'], shape=(2067,), dtype=object)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import balanced_accuracy_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "class BenchmarkClassifier():\n",
    "\n",
    "    def __init__(self, feature_col:str='query_length'):\n",
    "\n",
    "        self.model = LogisticRegression(class_weight='balanced', C=1) # , multi_class='multinomial') \n",
    "        self.scaler = StandardScaler()\n",
    "        self.feature_col = feature_col\n",
    "\n",
    "    def predict(self, df:pd.DataFrame, include_outputs:bool=True):\n",
    "        X = df[[self.feature_col]].values\n",
    "        X = self.scaler.transform(X)\n",
    "        model_labels = self.model.predict(X)\n",
    "        model_outputs = self.model.predict_proba(X)\n",
    "\n",
    "        return model_labels if (not include_outputs) else (model_labels, model_outputs)\n",
    "\n",
    "    def fit(self, df:pd.DataFrame):\n",
    "        X, labels = df[[self.feature_col]].values, df['label'].values\n",
    "        X = self.scaler.fit_transform(X)\n",
    "        self.model.fit(X, labels)\n",
    "\n",
    "model = BenchmarkClassifier()\n",
    "model.fit(train_df)\n",
    "\n",
    "df = test_df[['label', 'query_length']].copy()\n",
    "df['model_label'], outputs = model.predict(test_df, include_outputs=True)\n",
    "df['model_output_0'], df['model_output_1'] = outputs[:, 0].ravel(), outputs[:, 1].ravel()\n",
    "\n",
    "print('Balanced accuracy:', balanced_accuracy_score(df.label.values, df.model_label.values))\n",
    "print('Precision (class 0):', precision(df, class_=0))\n",
    "print('Recall (class 0):', recall(df, class_=0))\n",
    "print('Number of false negatives:', ((df.label == 1) & (df.model_label == 0)).sum())\n",
    "\n",
    "\n",
    "confusion_matrix = np.where((df.model_label == 1) & (df.label == 0), 'false positive', '')\n",
    "confusion_matrix = np.where((df.model_label  == 1) & (df.label == 1), 'true positive', confusion_matrix)\n",
    "confusion_matrix = np.where((df.model_label == 0) & (df.label == 1), 'false negative', confusion_matrix)\n",
    "confusion_matrix = np.where((df.model_label  == 0) & (df.label == 0), 'true negative', confusion_matrix)\n",
    "df['confusion_matrix'] = confusion_matrix\n",
    "\n",
    "benchmark_fn = df[df.confusion_matrix == 'false negative'].index\n",
    "fn = pred_df[pred_df.confusion_matrix == 'false negative'].index\n",
    "np.intersect1d(fn, benchmark_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There still seems to be something wrong with the training data, perhaps I am still misclassifying the spurious sequences?\n",
    "# It is possible my \"in-frame\" check is wrong, it's possible I should only see if either edge is a multiple of 3 base pairs away, \n",
    "# as opposed to both.\n",
    "\n",
    "# It seems as though 684 of the sequences flagged as spurious have an overlap length of 0, so maybe I should be counting those as intergenic?\n",
    "# Ok, so I think because of how frequent gene overlap is, I should be allowing some (considering genes with less than a certain amount of overlap to be intergenic)\n",
    "# I should also be very conservative, i.e. allow a comparatively large amount of overlap. \n",
    "# https://pmc.ncbi.nlm.nih.gov/articles/PMC525685/ \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Curious about the distribution of overlap sizes... \n",
    "\n",
    "\n",
    "# def figure(ref_df:pd.DataFrame, path:str=None, min_overlap_fraction:float=0.8):\n",
    "#     # ref_df = load_ref(genome_ids=genome_ids)\n",
    "#     ref_df = ref_df[ref_df.spurious | ref_df.real].copy()\n",
    "\n",
    "#     fig = plt.figure(figsize=(10, 10), layout='tight')\n",
    "#     grid = GridSpec(nrows=1, ncols=1, figure=fig)\n",
    "#     ax_a = fig.add_subplot(grid[0, 0])\n",
    "\n",
    "#     ax_a_df = pd.DataFrame(index=ref_df.index)\n",
    "#     ax_a_df['length'] = get_lengths(ref_df, top_hit=False, units='nt')\n",
    "#     ax_a_df['overlap_length'] = ref_df['top_hit_overlap_length']\n",
    "#     ax_a_df['overlap_fraction'] = ax_a_df.overlap_length / ax_a_df.length # This is relative to the query sequence. \n",
    "#     ax_a_df['label'] = np.select([ref_df.real, ref_df.spurious], ['real', 'spurious'], default='none')\n",
    "#     if min_overlap_fraction is not None:\n",
    "#         ax_a_df = ax_a_df[ax_a_df.overlap_fraction > min_overlap_fraction].copy()\n",
    "\n",
    "#     colors = ['darkseagreen', 'indianred']\n",
    "#     # sns.kdeplot(data=ax_a_df, ax=ax_a, hue='label', hue_order=['real', 'spurious'], palette=colors, common_norm=False, x='overlap_fraction')\n",
    "#     sns.ecdfplot(data=ax_a_df, ax=ax_a, hue='label', hue_order=['real', 'spurious'], palette=colors, x='overlap_fraction')\n",
    "\n",
    "#     plt.show()\n",
    "#     return ax_a_df\n",
    "\n",
    "\n",
    "# ax_a_df = figure(ref_df, min_overlap_fraction=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tripy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
